---
title: "Benchmarking"
output: 
  rmarkdown::html_document:
    theme: flatly
    toc: true
    toc_float: true
    toc_depth: 5
vignette: >
  %\VignetteIndexEntry{benchmarking}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\VignetteBuild{true}
---

```{r set_wd, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
dir <- "/Users/liuxin/macpie_Dev/"
devtools::load_all(paste0(dir, "macpie/"))

```


```{r setup}
library(macpie)
library(tibble)
library(stringr)
library(pheatmap)
library(ggiraph)
library(tidyseurat)
library(purrr)
library(DT)
library(ggrepel)
library(ggalluvial) 
library(ggplot2)
library(ggvenn)
library(limma)
```


## DRUGseq results


```{r}
# batch24 <- readRDS(paste0(dir,"DRUGseqData/Exp_batch24.Rds"))
```

```{r}
# load(paste0(dir,"DRUGseqData/de.RData")) 
# batch24_de <- de_res%>%filter(batch_id=="24")
# saveRDS(batch24_de, file = paste0(dir,"DRUGseqData/DE_batch24.Rds"))
batch24_de <- readRDS(paste0(dir,"DRUGseqData/DE_batch24.Rds"))
head(batch24_de)

```

```{r}

FF86_res <- batch24_de %>% filter(cmpd_sample_id=="FF-86-NH56")
ff86_res_toptable <- FF86_res[,13:18]
ff86_res_toptable <- ff86_res_toptable %>% 
  separate(gene.ID, into = c("gene", "chrom"), sep = ",") %>%
  select(-chrom) %>% mutate(combined_id ="FF_86_NH56_10") %>%
  rename(log2FC=logFC, p_value_adj = adj.P.Val)

```


```{r}
head(ff86_res_toptable)
```

```{r, fig.height=6}
plot_volcano(ff86_res_toptable, max.overlaps =5)
```

```{r}
ff86_res_toptable %>% filter(p_value_adj <0.05 & log2FC >0) %>% nrow()

drugseq_deg <- ff86_res_toptable %>% filter(p_value_adj <0.05 & log2FC >0) %>% select(gene) %>% pull()
```

```{r}
#show histogram of p values
ggplot(ff86_res_toptable, aes(x = P.Value)) +
  geom_histogram(binwidth = 0.01, fill = "blue", color = "black", alpha = 0.7) +
  labs(title = "Histogram of Adjusted P-values", x = "Adjusted P-value", y = "Frequency") +
  theme_minimal()
```



## macpie results

## All DMSO wells

### Load filered data


```{r filter_genes_batch24}
# mac_filtered <- filter_genes_by_expression(as_mac,
#                 group_by = "combined_id", min_counts = 10,
#                 min_samples = min_sample_num)

# saveRDS(mac_filtered,
# file = paste0(dir, "DRUGseqData/macpie_filtered_batch24.Rds"))


mac_filtered <- readRDS(paste0(dir, "/DRUGseqData/macpie_filtered_batch24.Rds"))
```


 


```{r violin_plots_batch24, fig.width=10, fig.height=6}
mac_filtered[["percent.mt"]] <- PercentageFeatureSet(mac_filtered, pattern = "^mt-|^MT-")
mac_filtered[["percent.ribo"]] <- PercentageFeatureSet(mac_filtered, pattern = "^Rp[slp][[:digit:]]|^Rpsa|^RP[SLP][[:digit:]]|^RPSA")
```


```{r}
mac_filtered$combined_id <- str_replace_all(mac_filtered$combined_id, "-","_")
```


#### Correction of batch effect 

According to DRUGseq metadata:

  - Wells with water are labelled as EC-27-RY89

  - Wells with DMSO are labelled as CB-43-EP73


<!-- # ```{r rle_plot_DMSO_batch24, fig.width=10, fig.height=6} -->
<!-- # # mac_filtered_cp <- mac_filtered %>% filter(combined_id %in% c("CB_43_EP73_0","FF_86_NH56_10")) -->
<!-- # mac_filtered_cp <- mac_filtered %>% filter(combined_id %in% c("CB_43_EP73_0")) -->
<!-- # plot_rle(mac_filtered_cp, label_column = "orig.ident", normalisation = "raw")+ scale_x_discrete(drop = FALSE) +  -->
<!-- #   theme(axis.text.x = element_text(angle = 45, hjust = 1)) -->
<!-- # plot_rle(mac_filtered_cp, label_column = "orig.ident", normalisation = "limma_voom")+ scale_x_discrete(drop = FALSE) +  -->
<!-- #   theme(axis.text.x = element_text(angle = 45, hjust = 1)) -->
<!-- # plot_rle(mac_filtered_cp, label_column = "orig.ident", normalisation = "DESeq2")+ scale_x_discrete(drop = FALSE) +  -->
<!-- #   theme(axis.text.x = element_text(angle = 45, hjust = 1)) -->
<!-- # plot_rle(mac_filtered_cp, label_column = "orig.ident", normalisation = "edgeR")+ scale_x_discrete(drop = FALSE) +  -->
<!-- #   theme(axis.text.x = element_text(angle = 45, hjust = 1)) -->
<!-- # ``` -->








*Note:* instead of discussing which correction methods we should use for this data, we only show the ways we detected and corrected batch effect here. As batch effect adjustment for sequencing data has been implemented in different methods, such as DESeq2, RUVSeq, edgeR. We highly recommend users thoroughly checking any batch effects and exploring different methods. 

In the next part of the vignette, we demonstrate a `batch` parameter has implemented in the differential expression test for batch correction. 



```{r}
##Limma trend 

# data <- mac_badDSMOremoved
# treatment_samples <- "FF_86_NH56_10"
# control_samples <- "CB_43_EP73_0"

 limma.trend<- function(data = data, treatment_samples, control_samples){
   data <- data[, grepl(paste0(treatment_samples, "|", control_samples), data$combined_id)]
   batch <- data$orig.ident
if (length(unique(data$combined_id)) < 2) {
      stop("Insufficient factors for differential expression analysis.")
    }
pheno_data <- data.frame(batch = as.factor(batch), condition = as.factor(data$combined_id))
combined_id <- data$combined_id
model_matrix <- if (length(batch) == 1) model.matrix(~0 + combined_id) else
      model.matrix(~0+combined_id + batch)

    dge <- DGEList(counts = data@assays$RNA$counts,
                   samples = pheno_data$condition,
                   group = pheno_data$condition)
    dge <- calcNormFactors(dge, method="TMMwsp")
    # keep <- edgeR::filterByExpr(dge, design = model_matrix)
    # dge  <- dge[keep, , keep.lib.sizes = FALSE]
    # clean.TMM<-log2(edgeR::cpm(dge,  normalized.lib.sizes=T,log=F)+1) 
    clean.TMM <- cpm(dge, log=TRUE, prior.count=3)
    myargs <- list(paste0("combined_id",
                          treatment_samples, "-",
                          paste0("combined_id", control_samples)),
                   levels = model_matrix)
    contrasts <- do.call(makeContrasts, myargs)
    
    lmfit <- lmFit(clean.TMM, model_matrix)
    lmfit <- contrasts.fit(lmfit, contrasts)
    lmfit <- eBayes(lmfit, trend = TRUE)
    #Warning: Zero sample variances detected, have been offset away from zero
    top_table<- topTable(lmfit, number = nrow(data)) %>%
      as.data.frame() %>%
      select("logFC", "t", "P.Value", "adj.P.Val") %>%
      rename("log2FC" = "logFC", "metric" = "t", "p_value" = "P.Value", "p_value_adj" = "adj.P.Val")%>%
      rownames_to_column("gene")

    return(top_table)}

    # top_table%>%filter(p_value_adj < 0.05 & log2FC>0)
```


<!-- # ```{r} -->
<!-- # limma_trend_safe <- function(seu, treatment, control) { -->
<!-- #   # subset Seurat -->
<!-- #   seu <- mac_filtered -->
<!-- #   # seu <- mac_badDSMOremoved -->
<!-- #   treatment <- "FF_86_NH56_10" -->
<!-- #   control <- "CB_43_EP73_0" -->
<!-- #   keep_cols <- grepl(paste0("^(?:", paste(c(treatment, control), collapse="|"), ")$"), -->
<!-- #                      seu$combined_id) -->
<!-- #   seu <- seu[, keep_cols] -->
<!-- #   # seu <- filter_genes_by_expression(seu, -->
<!-- #   #                                group_by = "orig.ident", -->
<!-- #   #                                min_counts = 1, -->
<!-- #   #                                min_samples = 3) -->
<!-- #   combined_id <- droplevels(factor(seu$combined_id)) -->
<!-- #   batch       <- droplevels(factor(seu$orig.ident)) -->
<!-- #  -->
<!-- #   # Build full model then drop batch if rank-deficient -->
<!-- #   mm_full <- model.matrix(~ 0 + combined_id + batch) -->
<!-- #   if (qr(mm_full)$rank < ncol(mm_full)){ -->
<!-- #     mm <- model.matrix(~ 0 + combined_id) -->
<!-- #   } else {mm <- mm_full} -->
<!-- #   colnames(mm) <- sub("^combined_id", "", colnames(mm)) -->
<!-- #  -->
<!-- #   # counts → DGEList → filterByExpr -->
<!-- #   counts <- seu@assays$RNA$counts -->
<!-- #   dge <- edgeR::DGEList(counts) -->
<!-- #   dge <- edgeR::calcNormFactors(dge, method = "TMMwsp") -->
<!-- #   keep <- edgeR::filterByExpr(dge, design = mm) -->
<!-- #   dge  <- dge[keep, , keep.lib.sizes = FALSE] -->
<!-- #  -->
<!-- #   # logCPM for limma-trend + drop constant rows -->
<!-- #   E <- edgeR::cpm(dge, log = TRUE, prior.count = 0.5, normalized.lib.sizes = TRUE) -->
<!-- #   E <- E[matrixStats::rowSds(as.matrix(E)) > 0, , drop = FALSE] -->
<!-- #   # sdv <- matrixStats::rowSds(as.matrix(E)) -->
<!-- #   # E <- E[sdv > summary(sdv)[4], , drop = FALSE] -->
<!-- #  -->
<!-- #   # estimable contrast name (assumes two levels remain) -->
<!-- #   # lev <- colnames(mm)[grepl(paste0("^", c(control, treatment), "$"), colnames(mm))] -->
<!-- #   # stopifnot(all(c(control, treatment) %in% lev)) -->
<!-- #   contr <- paste0(treatment, "-", control) -->
<!-- #   Cmat  <- limma::makeContrasts(contrasts = contr, levels = mm) -->
<!-- #  -->
<!-- #   fit <- limma::lmFit(E, mm) -->
<!-- #   fit <- limma::contrasts.fit(fit, Cmat) -->
<!-- #   fit <- limma::eBayes(fit, trend = TRUE) -->
<!-- #   fit <- limma::eBayes(limma::contrasts.fit(limma::lmFit(E, mm), Cmat), trend=TRUE) -->
<!-- #   fit_treat <- limma::treat(fit, lfc = log2(1.2))  # e.g., 20% FC -->
<!-- #   out <- limma::topTreat(fit_treat, number=Inf) -->
<!-- #  -->
<!-- #  -->
<!-- #   # out <- limma::topTable(fit, number = Inf) -->
<!-- #   out <- transform(out, -->
<!-- #                    gene = rownames(out), -->
<!-- #                    log2FC = logFC, -->
<!-- #                    metric = t, -->
<!-- #                    p_value = P.Value, -->
<!-- #                    p_value_adj = adj.P.Val)[, c("gene","log2FC","metric","p_value","p_value_adj")] -->
<!-- #   rownames(out) <- NULL -->
<!-- #   out -->
<!-- # } -->
<!-- #  -->
<!-- # #volcano plot -->
<!-- # limma_trend_safe_genes <- out%>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull() -->
<!-- # length(limma_trend_safe_genes) -->
<!-- # length(intersect(limma_trend_safe_genes, drugseq_deg)) -->
<!-- #  -->
<!-- # out <- out %>% -->
<!-- #     mutate(in_drugseq_deg = ifelse(gene %in% intersect(limma_trend_safe_genes, drugseq_deg), "yes", "no")) -->
<!-- # ggplot(out, aes(x = log2FC, y = -log10(p_value_adj), color = in_drugseq_deg)) + -->
<!-- #     geom_point() + -->
<!-- #     scale_color_manual(values = c("no"="#bdbdbd","yes"="#2b8cbe"))+ -->
<!-- #     theme_classic() -->
<!-- # ``` -->




<!-- # ```{r} -->
<!-- # dmso<- subset(mac_filtered, subset=combined_id=="CB_43_EP73_0" & orig.ident == "VH02012944") -->
<!-- # counts_d <- dmso@assays$RNA$counts -->
<!-- # # make human-friendly WELL IDs for output/labels -->
<!-- # well_colnames <- paste0(dmso$orig.ident, "_", dmso$Well_ID) -->
<!-- # names(well_colnames) <- rownames(dmso@meta.data) -->
<!-- # #rename the colnames of counts_d -->
<!-- # colnames(counts_d) <- well_colnames[colnames(counts_d)] -->
<!-- #  -->
<!-- # cpm_filter <- 1 -->
<!-- # y <- edgeR::DGEList(counts_d, group = dmso$orig.ident) -->
<!-- # keep <- rowSums(edgeR::cpm(y) > cpm_filter) >= 16  -->
<!-- # y <- y[keep,, keep.lib.sizes = FALSE] -->
<!-- #  -->
<!-- # # TMMwsp normalization and CPM calculation -->
<!-- # y <- edgeR::calcNormFactors(y, method = "TMMwsp") -->
<!-- # log_cpm_tmm <- cpm(y, log=TRUE, normalized.lib.sizes = TRUE) -->
<!-- #  -->
<!-- # #plot cpm in boxplot -->
<!-- # #covert to long format for ggplot -->
<!-- # df_long <- as.data.frame(log_cpm_tmm) %>% -->
<!-- #   rownames_to_column(var = "gene") %>% -->
<!-- #   pivot_longer(-gene, names_to = "sample", values_to = "log_cpm") -->
<!-- #  -->
<!-- # #plot ggplotboxplot -->
<!-- # ggplot(df_long, aes(x = sample, y = log_cpm)) + -->
<!-- #   geom_boxplot(outlier.size = 0.5) + -->
<!-- #   labs(x = "Sample", y = "Log2 CPM", title = "Boxplot of Log2 CPM after TMMwsp normalization") + -->
<!-- #   theme_classic() + -->
<!-- #   theme(axis.text.x = element_text(angle = 45, hjust = 1)) -->
<!-- #  -->
<!-- # cors_pear <- cor(log_cpm_tmm, method = "pearson") -->
<!-- # pheatmap(cors_pear, main = "Sample–sample correlation (Pearson, log2-CPM)") -->
<!-- #  -->
<!-- # # Spearman - expects outliers -->
<!-- # cors_spear <- cor(log_cpm_tmm, method = "spearman") -->
<!-- # pheatmap(cors_spear, main = "Sample–sample correlation (Spearman, log2-CPM)") -->
<!-- # #rowmeans then rank on the rowmeans -->
<!-- # R <- cor(log_cpm_tmm, method = "spearman", use = "pairwise.complete.obs") -->
<!-- # diag(R) <- NA -->
<!-- #  -->
<!-- # # Fisher z-average is the right way to average correlations -->
<!-- # Z <- atanh(pmin(pmax(R, -0.999999), 0.999999))   # clip to avoid Inf -->
<!-- # score_z <- rowMeans(Z, na.rm = TRUE)               # per-sample avg to all others -->
<!-- # score_r <- tanh(score_z)                           # back to r -->
<!-- #  -->
<!-- # # Rank and take top 5 -->
<!-- # top5 <- sort(score_r, decreasing = TRUE)[1:5] -->
<!-- # top5 -->
<!-- # ``` -->



<!-- # ```{r} -->
<!-- # top_table %>% filter(p_value_adj < 0.05 & log2FC>0) %>% nrow() -->
<!-- # alldmso_degs <- top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull() -->
<!-- # ``` -->



<!-- # ```{r} -->
<!-- # length(intersect(alldmso_degs, drugseq_deg)) -->
<!-- # ``` -->



<!-- # ```{r} -->
<!-- # top_table <- top_table %>% -->
<!-- #   arrange(p_value_adj, desc(log2FC)) %>% -->
<!-- #   mutate(gene = factor(gene, levels = unique(gene))) -->
<!-- # # add a column if there are in the intersect(alldmso_degs, drugseq_deg) -->
<!-- # top_table <- top_table %>% -->
<!-- #   mutate(in_drugseq_deg = ifelse(gene %in% intersect(alldmso_degs, drugseq_deg), "yes", "no")) -->
<!-- # ``` -->
<!-- #  -->
<!-- # ```{r} -->
<!-- # plt_ecdf <- ggplot(top_table, aes(x = p_value_adj, color = in_drugseq_deg)) + -->
<!-- #   stat_ecdf(size = 1) + -->
<!-- #   scale_x_continuous(trans = "log10", breaks = c(1e-6,1e-4,5e-2)) + -->
<!-- #   labs(x = "Adjusted p-value (log10 scale)", y = "ECDF", -->
<!-- #        color = "In DRUGseq DEGs") + -->
<!-- #   theme_classic() -->
<!-- # plt_ecdf -->
<!-- # ``` -->
<!-- KS test (padj) -->

<!-- ```{r} -->
<!-- ks.test(top_table$p_value_adj[top_table$in_drugseq_deg=="yes"], -->
<!--         top_table$p_value_adj[top_table$in_drugseq_deg=="no"], alternative = "greater") -->

<!-- ``` -->
<!-- # ```{r} -->
<!-- # ggplot(top_table, aes(x = log2FC, y = -log10(p_value_adj), color = in_drugseq_deg)) + -->
<!-- #   geom_point(alpha = 0.6, size = 1.2) + -->
<!-- #   geom_hline(yintercept = -log10(0.05), linetype = "dashed") + -->
<!-- #   geom_vline(xintercept = c(-1, 1), linetype = "dashed") + -->
<!-- #   scale_color_manual(values = c("no"="#bdbdbd","yes"="#2b8cbe")) + -->
<!-- #   labs(x = "log2FC", y = expression(-log[10]("adj p")), color = "In DRUGseq DEGs") + -->
<!-- #   theme_classic() -->
<!-- #  -->
<!-- # # label a few top overlapping genes -->
<!-- # lab_genes <- top_table[top_table$in_drugseq_deg=="yes", ] |> -->
<!-- #   dplyr::arrange(p_value_adj, dplyr::desc(log2FC)) |> -->
<!-- #   head(15) -->
<!-- #  -->
<!-- # ggplot(top_table, aes(x = log2FC, y = -log10(p_value_adj), color = in_drugseq_deg)) + -->
<!-- #   geom_point(alpha = 0.6, size = 1.2) + -->
<!-- #   geom_text_repel(data = lab_genes, aes(label = gene), size = 3, max.overlaps = 50) + -->
<!-- #   scale_color_manual(values = c("no"="#bdbdbd","yes"="#2b8cbe"))+ -->
<!-- #   theme_classic() -->
<!-- # ``` -->

<!-- # ```{r} -->
<!-- # ks <- c(25,50,100,200,500,1000) -->
<!-- # prec_at_k <- sapply(ks, function(k) -->
<!-- #   mean(top_table$in_drugseq_deg[1:k] == "yes") -->
<!-- # ) -->
<!-- # plot(ks, prec_at_k, type = "b", xlab = "k (top-k by macpie)", -->
<!-- #      ylab = "Precision@k (fraction in DRUGseq set)") -->
<!-- #  -->
<!-- # ``` -->

<!-- # ```{r} -->
<!-- # library(dplyr) -->
<!-- # library(ggplot2) -->
<!-- # library(ggalluvial) -->
<!-- #  -->
<!-- # macpie_tbl <- top_table %>% filter(p_value_adj < 0.05 & log2FC>0) -->
<!-- # drugseq_tbl <- ff86_res_toptable %>% filter(p_value_adj < 0.05 & log2FC>0) -->
<!-- #  -->
<!-- #  -->
<!-- # # macpie_tbl: gene, p_value_adj, log2FC -->
<!-- # # drugseq_tbl: gene, p_value_adj, log2FC -->
<!-- #  -->
<!-- # df <- macpie_tbl %>% -->
<!-- #   transmute(gene, p_value_adj_mac = p_value_adj, log2FC_mac = log2FC) %>% -->
<!-- #   inner_join( -->
<!-- #     drugseq_tbl %>% transmute(gene, p_value_adj_drg = p_value_adj, log2FC_drg = log2FC), -->
<!-- #     by = "gene" -->
<!-- #   ) -->
<!-- #  -->
<!-- # # Rank within each method by (padj, |log2FC|), tie-safe -->
<!-- # rank_by_two <- function(padj, lfc) { -->
<!-- #   ord <- order(padj, -abs(lfc), na.last = TRUE) -->
<!-- #   r   <- integer(length(padj)) -->
<!-- #   r[ord] <- seq_along(ord) -->
<!-- #   r -->
<!-- # } -->
<!-- # df <- df %>% -->
<!-- #   mutate( -->
<!-- #     rank_mac = rank_by_two(p_value_adj_mac, log2FC_mac), -->
<!-- #     rank_drg = rank_by_two(p_value_adj_drg, log2FC_drg) -->
<!-- #   ) -->
<!-- #  -->
<!-- # # Define rank bins -->
<!-- # cuts <- c(0, 25, 50, 100, 200, 300, 500, Inf) -->
<!-- # labs <- c("Top25","26–50","51–100","101–200","201–300", "301-500",">500") -->
<!-- #  -->
<!-- # df <- df %>% -->
<!-- #   mutate( -->
<!-- #     mac_bin = cut(rank_mac, breaks = cuts, labels = labs, right = TRUE, include.lowest = TRUE), -->
<!-- #     drg_bin = cut(rank_drg, breaks = cuts, labels = labs, right = TRUE, include.lowest = TRUE) -->
<!-- #   ) -->
<!-- #  -->
<!-- # # ranking bin changes -->
<!-- # bin_index <- function(x) match(x, labs)   # lower index = "more top" -->
<!-- # df <- df %>% -->
<!-- #   mutate( -->
<!-- #     mac_idx = bin_index(mac_bin), -->
<!-- #     drg_idx = bin_index(drg_bin), -->
<!-- #     movement = case_when( -->
<!-- #       is.na(mac_idx) | is.na(drg_idx) ~ NA_character_, -->
<!-- #       mac_idx < drg_idx ~ "Higher rank in macpie",   # moved up into a more-top bin -->
<!-- #       mac_idx > drg_idx ~ "Lower rank in macpie",    # moved down -->
<!-- #       TRUE               ~ "Same rank" -->
<!-- #     ), -->
<!-- #     movement = factor(movement, levels = c("Higher rank in macpie","Same rank","Lower rank in macpie")) -->
<!-- #   ) %>% -->
<!-- #   tidyr::drop_na(mac_bin, drg_bin, movement) -->
<!-- #  -->
<!-- # # Aggregate flows for alluvial -->
<!-- # flows <- df %>% -->
<!-- #   count(mac_bin, drg_bin, movement, name = "n") %>% -->
<!-- #   mutate( -->
<!-- #     mac_bin = factor(mac_bin, levels = labs), -->
<!-- #     drg_bin = factor(drg_bin, levels = labs) -->
<!-- #   ) -->
<!-- #   -->
<!-- # ggplot(flows, aes(axis1 = mac_bin, axis2 = drg_bin, y = n)) + -->
<!-- #   geom_alluvium(aes(fill = movement), alpha = 0.75, width = 0.14, knot.pos = 0.4) + -->
<!-- #   geom_stratum(width = 0.14, color = "grey30") + -->
<!-- #   geom_text(stat = "stratum", aes(label = after_stat(stratum)), size = 3) + -->
<!-- #   scale_x_discrete(limits = c("Macpie rank", "DRUGseq rank"), expand = c(.05, .05)) + -->
<!-- #   scale_fill_manual(values = c("Higher rank in macpie" = "orange", -->
<!-- #                                "Same rank"        = "#7570b3", -->
<!-- #                                "Lower rank in macpie" = "grey60")) + -->
<!-- #   labs(x = NULL, y = "Gene count", -->
<!-- #        fill = "Movement vs DRUGseq", -->
<!-- #        subtitle = "Only DEGs with padj < 0.05 & log2FC>0 (both methods)") + -->
<!-- #   theme_classic() -->
<!-- #  -->
<!-- # ``` -->


```{r}
# make a function of alluvial plot with ggalluvial to compare the ranking of DEGs from macpie and DRUGseq
plot_alluvial <- function(macpie_tbl, drugseq_tbl){
  df <- macpie_tbl %>%
  transmute(gene, p_value_adj_mac = p_value_adj, log2FC_mac = log2FC) %>%
  inner_join(
    drugseq_tbl %>% transmute(gene, p_value_adj_drg = p_value_adj, log2FC_drg = log2FC),
    by = "gene"
  )

#  Rank within each method by (padj, |log2FC|), tie-safe
rank_by_two <- function(padj, lfc) {
  ord <- order(padj, -abs(lfc), na.last = TRUE)
  r   <- integer(length(padj))
  r[ord] <- seq_along(ord)
  r
}
df <- df %>%
  mutate(
    rank_mac = rank_by_two(p_value_adj_mac, log2FC_mac),
    rank_drg = rank_by_two(p_value_adj_drg, log2FC_drg)
  )

# Define rank bins
cuts <- c(0, 25, 50, 100, 200, 300, 500, Inf)
labs <- c("Top25","26–50","51–100","101–200","201–300", "301-500",">500")

df <- df %>%
  mutate(
    mac_bin = cut(rank_mac, breaks = cuts, labels = labs, right = TRUE, include.lowest = TRUE),
    drg_bin = cut(rank_drg, breaks = cuts, labels = labs, right = TRUE, include.lowest = TRUE)
  )

# Ranking movement category (color)
bin_index <- function(x) match(x, labs)   # lower index = "more top"
df <- df %>%
  mutate(
    mac_idx = bin_index(mac_bin),
    drg_idx = bin_index(drg_bin),
    movement = case_when(
      is.na(mac_idx) | is.na(drg_idx) ~ NA_character_,
      mac_idx < drg_idx ~ "Higher rank in macpie",   # moved up into a more-top bin
      mac_idx > drg_idx ~ "Lower rank in macpie",    # moved down
      TRUE               ~ "Same rank"
    ),
    movement = factor(movement, levels = c("Higher rank in macpie","Same rank","Lower rank in macpie"))
  ) %>%
  tidyr::drop_na(mac_bin, drg_bin, movement)

#  Aggregate flows for alluvial
flows <- df %>%
  count(mac_bin, drg_bin, movement, name = "n") %>%
  mutate(
    mac_bin = factor(mac_bin, levels = labs),
    drg_bin = factor(drg_bin, levels = labs)
  )

ggplot(flows, aes(axis1 = mac_bin, axis2 = drg_bin, y = n)) +
  geom_alluvium(aes(fill = movement), alpha = 0.75, width = 0.14, knot.pos = 0.4) +
  geom_stratum(width = 0.14, color = "grey30") +
  geom_text(stat = "stratum", aes(label = after_stat(stratum)), size = 3) +
  scale_x_discrete(limits = c("Macpie rank", "DRUGseq rank"), expand = c(.05, .05)) +
  scale_fill_manual(values = c("Higher rank in macpie" = "orange",
                               "Same rank"        = "#7570b3",
                               "Lower rank in macpie" = "grey60")) +
  labs(x = NULL, y = "Gene count",
       fill = "Movement vs DRUGseq",
       subtitle = "Only DEGs with padj < 0.05 & log2FC>0 (both methods)") +
  theme_classic()
  
}
```


####  Differential gene expression 


In here, you can specify a single condition in the combined_id column and compare with DMSO (i.e.CB_43_EP73_0). By using the plate IDs in the column of orig.ident as the input for batch parameter, `compute_singe_de` function can perform differential expression analysis using the preferred method (limma voom in this example) with batch information. 



```{r}
methods <- c("limma_voom", "DESeq2", "edgeR", "limma_trend")

methods_res <- lapply(methods, function(m){
  
  message("\n","Processing method: ", m,"\n") 
  # m<-"limma_voom"
  treatment_samples <- "FF_86_NH56_10"
  control_samples <- "CB_43_EP73_0"
  subset <- mac_filtered%>%filter(combined_id%in%c(treatment_samples,control_samples))

  batch <- subset$orig.ident
  if (m!="limma_trend"){
    subset <- filter_genes_by_expression(subset,
                           group_by = "combined_id", min_counts = 5,
                           min_samples = 1)
    top_table <- compute_single_de(subset, treatment_samples, control_samples, method = m, batch = batch)
  } else{
    top_table <- limma.trend(data = subset, treatment_samples = treatment_samples, control_samples = control_samples)
  }
  
  # plot(plot_volcano(top_table, max.overlaps = 5))
  alldmso_degs <- top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull()
  length(intersect(alldmso_degs, drugseq_deg))
  
  top_table <- top_table %>%
    arrange(p_value_adj, desc(log2FC)) %>%
    mutate(gene = factor(gene, levels = unique(gene)))
  # add a column if there are in the intersect(alldmso_degs, drugseq_deg)
  top_table <- top_table %>%
    mutate(in_drugseq_deg = ifelse(gene %in% intersect(alldmso_degs, drugseq_deg), "yes", "no"))
  
  plt_ecdf <- ggplot(top_table, aes(x = p_value_adj, color = in_drugseq_deg)) +
    stat_ecdf(size = 1) +
    scale_x_continuous(trans = "log10", breaks = c(1e-6,1e-4,5e-2)) +
    labs(x = "Adjusted p-value (log10 scale)", y = "ECDF",
         color = "In DRUGseq DEGs") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    theme_classic()
  
  
  ks_test_results <- ks.test(top_table$p_value_adj[top_table$in_drugseq_deg=="yes"],
          top_table$p_value_adj[top_table$in_drugseq_deg=="no"], alternative = "greater")
  
  # ggplot(top_table, aes(x = log2FC, y = -log10(p_value_adj), color = in_drugseq_deg)) +
  #   geom_point(alpha = 0.6, size = 1.2) +
  #   geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  #   geom_vline(xintercept = c(-1, 1), linetype = "dashed") +
  #   scale_color_manual(values = c("no"="#bdbdbd","yes"="#2b8cbe")) +
  #   labs(x = "log2FC", y = expression(-log[10]("adj p")), color = "In DRUGseq DEGs") +
  #   theme_classic()
  # 
  # label a few top overlapping genes
  lab_genes <- top_table[top_table$in_drugseq_deg=="yes", ] |>
    dplyr::arrange(p_value_adj, dplyr::desc(log2FC)) 
  
  volcano_overlap <- ggplot(top_table, aes(x = log2FC, y = -log10(p_value_adj), color = in_drugseq_deg)) +
    geom_point(alpha = 0.6, size = 1.2) +
    geom_text_repel(data = lab_genes, aes(label = gene), size = 3, max.overlaps = 50) +
    scale_color_manual(values = c("no"="#bdbdbd","yes"="#2b8cbe"))+
    theme_classic()
  
  ks <- c(25,50,100,200,500,1000)
  prec_at_k <- sapply(ks, function(k)
    mean(top_table$in_drugseq_deg[1:k] == "yes")
  )
  # plot(ks, prec_at_k, type = "b", xlab = "k (top-k by macpie)",
  #      ylab = "Precision@k (fraction in DRUGseq set)")
  
  ks_plot <- ggplot(data.frame(k=ks, prec=prec_at_k), aes(x=k, y=prec)) +
    geom_point() + geom_line() +
    labs(x = "k (top-k by macpie)", y = "Precision@k (fraction in DRUGseq set)")+
    theme_classic()
  
  # alluvial plot
  macpie_tbl <- top_table %>% filter(p_value_adj < 0.05 & log2FC>0)
  drugseq_tbl <- ff86_res_toptable %>% filter(p_value_adj < 0.05 & log2FC>0)
  alluvial_plot <- plot_alluvial(macpie_tbl = macpie_tbl, drugseq_tbl = drugseq_tbl)
  
  #return 
  result_list <- list(top_table = top_table,
                      num_degs_macpie = length(alldmso_degs),
                      n_overlap = length(intersect(alldmso_degs, drugseq_deg)),
                      ecdf_plot = plt_ecdf,
                      ks_test_results = ks_test_results,
                      volcano_plot = volcano_overlap,
                      ks_plot = ks_plot,
                      alluvial_plot = alluvial_plot)
  return(result_list)
  
})

names(methods_res) <- methods

```


#### Summary table


```{r}
#get a table to show number of DEGs and number of overlapping genes with DRUGseq for each method
deg_summary <- map_df(methods_res, function(x) {
  data.frame(
    num_degs_macpie = x$num_degs_macpie,
    n_overlap = x$n_overlap,
    num_degs_DRUGseq = length(drugseq_deg)
  )
}, .id = paste0("macpie_methods"))

deg_summary
```




<!-- #### Common genes from methods in macpie -->

<!-- ```{r} -->

<!-- library(eulerr) -->
<!-- three_methods_degs <- list( -->
<!--   limma_voom = methods_res$limma_voom$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!--   DESeq2 = methods_res$DESeq2$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!--   edgeR = methods_res$edgeR$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!--   limma_trend = methods_res$limma_trend$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull() -->
<!-- ) -->


<!-- ggvenn( -->
<!--   three_methods_degs, -->
<!--   fill_color = macpie_colours$discrete,   # optional custom colors -->
<!--   stroke_size = 0.5,    # border thickness -->
<!--   set_name_size = 4,    # label size -->
<!--   text_size = 3         # count label size -->
<!-- ) +  -->
<!--   ggtitle("Genes overlap among three macpie methods") -->

<!-- fit <- euler(three_methods_degs) -->
<!-- plot(fit, quantities = TRUE, legend = TRUE)    -->

<!-- ``` -->


<!-- #### Common genes from three methods in macpie and DRUGseq -->

<!-- # ```{r} -->
<!-- # three_methods_degs <- list( -->
<!-- #   limma_voom = methods_res$limma_voom$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!-- #   DESeq2 = methods_res$DESeq2$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!-- #   edgeR = methods_res$edgeR$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!-- #   limma_trend = methods_res$limma_trend$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!-- #   DRUGseq = drugseq_deg -->
<!-- # ) -->
<!-- #  -->
<!-- #  -->
<!-- # ggvenn( -->
<!-- #   three_methods_degs, -->
<!-- #   fill_color = macpie_colours$discrete,   # optional custom colors -->
<!-- #   stroke_size = 0.5,    # border thickness -->
<!-- #   set_name_size = 4,    # label size -->
<!-- #   text_size = 3         # count label size -->
<!-- # ) + -->
<!-- #   ggtitle("Genes overlap among macpie methods and DRUGseq results") -->
<!-- #  -->
<!-- # fit <- euler(three_methods_degs) -->
<!-- # plot(fit, quantities = TRUE, legend = TRUE)     -->
<!-- # ``` -->


<!-- ### ECDF plot -->
<!-- ```{r} -->
<!-- methods_res$limma_voom$ecdf_plot -->
<!-- methods_res$DESeq2$ecdf_plot -->
<!-- methods_res$edgeR$ecdf_plot -->
<!-- ``` -->




### Overlapping volcano plot

```{r}
methods_res$limma_voom$volcano_plot
methods_res$DESeq2$volcano_plot
methods_res$edgeR$volcano_plot
methods_res$limma_trend$volcano_plot
```



<!-- ### Precision@k -->
<!-- ```{r} -->
<!-- methods_res$limma_voom$ks_plot -->
<!-- methods_res$DESeq2$ks_plot -->
<!-- methods_res$edgeR$ks_plot -->
<!-- methods_res$limma_trend$ks_plot -->
<!-- ``` -->



### Alluvial plot

```{r}
methods_res$limma_voom$alluvial_plot
methods_res$DESeq2$alluvial_plot
methods_res$edgeR$alluvial_plot
methods_res$limma_trend$alluvial_plot
```


```{r}
#append top_table from each method into a single data frame
all_methods_top_table <- map_df(methods_res, "top_table", .id = "method")
head(all_methods_top_table)

#if in_drugseq_deg is "no", change method to "DRUGseq"
all_methods_top_table <- all_methods_top_table %>%
  mutate(method = ifelse(in_drugseq_deg == "no", "not in DRUGseq", method))

#rename methods
all_methods_top_table$method <- recode(all_methods_top_table$method,
                                        "limma_voom" = "macpie:limma_voom",
                                        "DESeq2" = "macpie:DESeq2",
                                        "edgeR" = "macpie:edgeR",
                                        "limma_trend" = "macpie:limma_trend")


ggplot(all_methods_top_table, aes(x = p_value_adj, color = method)) +
    stat_ecdf(size = 1) +
    scale_x_continuous(trans = "log10", breaks = c(1e-6,1e-4,5e-2)) +
    labs(x = "Adjusted p-value (log10 scale)", y = "ECDF",
         color = "In DRUGseq DEGs") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    theme_classic()


```


```{r}
rm(top_table, alldmso_degs, deg_summary)
```




## Good DMSO - top 5 from macpie 

From public DRUG-seq analysis pipeline, authors identified two reference controls: all DMSO wells and the ‘good DMSO’ wells. 

We know these good DMSO wells for batch 24 from their published data:

  - VH02012942: I23, M23
  
  - VH02012944: D23, H23
  
  - VH02012956: I23, J23
  

```{r}
batch24 <- readRDS(paste0(dir,"DRUGseqData/Exp_batch24.Rds"))
names(batch24)
```  

```{r metdata}
#make a combined metadata for three plates
batch24_metadata <- batch24 %>% 
  map_dfr(~ {
    .x$Annotation %>%
      mutate(
        Plate_ID        = plate_barcode,
        Well_ID         = well_id,
        Barcode         = paste0(plate_barcode, "_", well_index),
        Row             = LETTERS[row],
        Column          = as.integer(col),
        Species         = "human",
        Cell_type       = "U2OS",
        Model_type      = "2D_adherent",
        Time            = as.factor(hours_post_treatment),
        Unit            = "h",
        Treatment_1     = cmpd_sample_id,
        Concentration_1 = as.numeric(concentration),
        Unit_1          = unit,
        Sample_type     = if_else(well_type == "SA" & col == 24,
                                  "Positive Control",
                                  well_type)
      )
  })


batch24_metadata <- batch24_metadata%>%select(-c(batch_id, plate_barcode,plate_index, well_id,
                                                 well_index, col, row, biosample_id, external_biosample_id,
                                                 cmpd_sample_id, well_type, cell_line_name, cell_line_ncn, concentration, unit, hours_post_treatment, Sample))
```



```{r count_matrix}
# create a combined UMI matrix for 3 plates
batch24_counts <- batch24 %>%
  map(~ {
    .x$UMI.counts %>%
      as.data.frame() %>% 
      rownames_to_column("gene_id") %>%
      separate(col = gene_id, into = c("gene_name", "chrom"), sep = ",") %>%
      mutate(gene_name = make.unique(gene_name)) %>%
      select(-chrom) %>%
      tibble::column_to_rownames(var = "gene_name") %>%
      as.matrix()
  })

binded_counts <- do.call(cbind, batch24_counts)


  
```



```{r load_data}
as_mac <- CreateSeuratObject(counts = binded_counts, 
  min.cells = 1, 
  min.features = 1)

as_mac<- as_mac%>% inner_join(batch24_metadata, by = c(".cell"="Barcode"))
```


Filtering steps

Include only good DMSO wells as controls

```{r}

as_mac$combined_id <- paste0(as_mac$Treatment_1,"_", as_mac$Concentration_1)

badDMSO <- as_mac@meta.data %>% filter(Treatment_1 == "CB-43-EP73") %>% 
 filter((Plate_ID == "VH02012942" & !(Well_ID %in% c("I23", "M23", "K23", "J23","C23"))) |
        (Plate_ID == "VH02012944" & !(Well_ID %in% c("I23", "M23", "J23", "G23", "K23")))|
        (Plate_ID == "VH02012956" & !  (Well_ID %in% c("I23", "J23", "O23","M23","K23"))))



keep_wells <- setdiff(rownames(as_mac@meta.data), rownames(badDMSO))


mac_badDSMOremoved <- as_mac[,keep_wells]

mac_badDSMOremoved$combined_id <- str_replace_all(mac_badDSMOremoved$combined_id, "-","_")


```


```{r}
min_sample_num <- min(table(mac_badDSMOremoved$combined_id))
mac_badDSMOremoved <- filter_genes_by_expression(mac_badDSMOremoved,
                                           group_by = "combined_id", min_counts =10,
                                           min_samples = min_sample_num)

```


```{r}
mac_badDSMOremoved[["percent.mt"]] <- PercentageFeatureSet(mac_badDSMOremoved, pattern = "^mt-|^MT-")
mac_badDSMOremoved[["percent.ribo"]] <- PercentageFeatureSet(mac_badDSMOremoved, pattern = "^Rp[slp][[:digit:]]|^Rpsa|^RP[SLP][[:digit:]]|^RPSA")

```

```{r}
p <- plot_plate_layout(mac_badDSMOremoved, "nCount_RNA", "combined_id") + facet_wrap(~orig.ident, ncol = 1) + 
  theme(strip.text = element_text(size=10),
        axis.text.x = element_text(size=10), 
        axis.text.y = element_text(size=8),
        legend.title = element_text(size=10),
        legend.text = element_text(size=8),
        trip.background = element_blank())
girafe(ggobj = p, 
  fonts = list(sans = "sans"),
  options = list(
    opts_hover(css = "stroke:black; stroke-width:1px;")
  ))
```




<!-- ```{r} -->
<!-- p <- plot_mds(mac_badDSMOremoved, group_by = "Sample_type", label = "combined_id", n_labels = 30) -->
<!-- #> tidyseurat says: Key columns are missing. A data frame is returned for independent data analysis. -->
<!-- #> tidyseurat says: Key columns are missing. A data frame is returned for independent data analysis. -->

<!-- p1 <- plot_mds(mac_badDSMOremoved, group_by = "orig.ident", label = "combined_id", n_labels = 30) -->
<!-- #> tidyseurat says: Key columns are missing. A data frame is returned for independent data analysis. -->
<!-- #> tidyseurat says: Key columns are missing. A data frame is returned for independent data analysis. -->


<!-- g <- patchwork::wrap_plots(list(p, p1), ncol = 1, nrow = 2, rel_widths = c(7, 7), rel_heights = c(10, 10)) -->
<!-- girafe( -->
<!--   ggobj      = g, -->
<!--   width_svg  = 10,    # 10 inches wide -->
<!--   height_svg = 10,     #  10 inches tall -->
<!--   fonts      = list(sans = "sans"), -->
<!--   options    = list(opts_hover(css = "stroke:black; stroke-width:0.8px;")) -->
<!-- ) -->
<!-- ``` -->


<!-- # ```{r} -->
<!-- # mac_sct <- SCTransform(mac_badDSMOremoved, verbose = FALSE) %>% -->
<!-- #     RunPCA(verbose = FALSE) %>% -->
<!-- #     RunUMAP(verbose = FALSE,dims = 1:30)  -->
<!-- #  -->
<!-- # DimPlot(mac_sct, reduction = "pca", group.by = "orig.ident", cols = macpie_colours$discrete) -->
<!-- # ``` -->





<!-- ```{r} -->
<!-- mac_badDSMOremoved_cp <- mac_badDSMOremoved %>% filter(combined_id %in% c("CB_43_EP73_0","FF_86_NH56_10")) -->
<!-- plot_rle(mac_badDSMOremoved_cp, label_column = "orig.ident", normalisation = "raw")+ scale_x_discrete(drop = FALSE) +    -->
<!--   theme(axis.text.x = element_text(angle = 45, hjust = 1)) -->
<!-- plot_rle(mac_badDSMOremoved_cp, label_column = "orig.ident", normalisation = "limma_voom")+ scale_x_discrete(drop = FALSE) +  -->
<!--   theme(axis.text.x = element_text(angle = 45, hjust = 1)) -->
<!-- plot_rle(mac_badDSMOremoved_cp, label_column = "orig.ident", normalisation = "DESeq2")+ scale_x_discrete(drop = FALSE) +  -->
<!--   theme(axis.text.x = element_text(angle = 45, hjust = 1)) -->
<!-- plot_rle(mac_badDSMOremoved_cp, label_column = "orig.ident", normalisation = "edgeR")+ scale_x_discrete(drop = FALSE) +  -->
<!--   theme(axis.text.x = element_text(angle = 45, hjust = 1)) -->
<!-- ``` -->







<!-- # ```{r} -->
<!-- # treatment_samples <- "FF_86_NH56_10" -->
<!-- # control_samples <- "CB_43_EP73_0" -->
<!-- # subset <- mac_badDSMOremoved%>%filter(combined_id%in%c(treatment_samples,control_samples)) -->
<!-- # subset <- filter_genes_by_expression(subset, -->
<!-- #                            group_by = "combined_id", min_counts = 5, -->
<!-- #                            min_samples = 2)  -->
<!-- # batch <- subset$orig.ident -->
<!-- # # badDMSO_out_top_table <- compute_single_de(mac_badDSMOremoved, treatment_samples, control_samples, method =  "DESeq2", batch = batch) -->
<!-- # # plot_volcano(badDMSO_out_top_table, max.overlaps = 6) -->
<!-- # pheno_data <- data.frame(batch = as.factor(batch), condition = as.factor(subset$combined_id)) -->
<!-- # combined_id <- subset$combined_id -->
<!-- # dds <- DESeqDataSetFromMatrix(countData = subset@assays$RNA$counts, -->
<!-- #                                   colData = pheno_data, -->
<!-- #                                   # design = ~ condition) -->
<!-- #                                   design = ~ condition + batch) -->
<!-- # # dds <- DESeq(dds,test = "LRT", reduced = ~ batch) -->
<!-- # dds <- DESeq(dds) -->
<!-- # res <- results(dds) -->
<!-- # #need to compare both condition and batch  -->
<!-- # # res <- results(dds, contrast = c("condition", treatment_samples, control_samples)) -->
<!-- #   top_table <- as.data.frame(res) %>% -->
<!-- #       select("log2FoldChange", "stat", "pvalue", "padj") %>% -->
<!-- #       rename("log2FC" = "log2FoldChange", "metric" = "stat", "p_value" = "pvalue", "p_value_adj" = "padj") %>% -->
<!-- #       rownames_to_column("gene") -->
<!-- # plot_volcano(top_table, max.overlaps = 6) -->
<!-- #  -->
<!-- # new_deseq2_degs<-top_table%>%filter(log2FC>0 & p_value_adj < 0.05) %>%select(gene)%>%pull() -->
<!-- # length(new_deseq2_degs) -->
<!-- # length(intersect(new_deseq2_degs, drugseq_deg)) -->
<!-- # ``` -->
<!-- #  -->
<!-- # ```{r} -->
<!-- # badDMSO_out_top_table %>% filter(p_value_adj <0.05& log2FC >1) %>% nrow() -->
<!-- #  -->
<!-- # macpie_deg <- badDMSO_out_top_table %>% filter(p_value_adj <0.05  & log2FC >1)%>%select(gene) %>% pull() -->
<!-- # ``` -->
<!-- #  -->
<!-- # ```{r} -->
<!-- # length(intersect(macpie_deg, drugseq_deg)) -->
<!-- # ``` -->


<!-- # ```{r} -->
<!-- # sub_mac <- subsample_genes(mac_badDSMOremoved, ngene = 3000, gselect = "random", seed = 1) -->
<!-- # sub_mac %>% nrow() -->
<!-- # zi_results <- check_zeroinflation(sub_mac,  -->
<!-- #                                   group_by = "combined_id", -->
<!-- #                                   samples = c("CB_43_EP73_0","FF_86_NH56_10"), -->
<!-- #                                   batch = 1, -->
<!-- #                                   cutoffs = c(0.1, 0.2)) -->
<!-- # zi_results$summary_by_group -->
<!-- # ``` -->





```{r}
methods <- c("limma_voom", "DESeq2", "edgeR", "limma_trend")

five_dmso_methods_res <- lapply(methods, function(m){
  message("Processing method: ", m) 
  # m<-"limma_voom"
  treatment_samples <- "FF_86_NH56_10"
  control_samples <- "CB_43_EP73_0"
  subset <- mac_badDSMOremoved%>%filter(combined_id%in%c(treatment_samples,control_samples))

  batch <- subset$orig.ident
  if (m!="limma_trend"){
      # subset <- filter_genes_by_expression(subset,
      #                        group_by = "combined_id", min_counts = 10,
      #                        min_samples = 1)
     badDMSO_out_top_table <- compute_single_de(subset, treatment_samples, control_samples, method =  m, batch = batch)
  } else{
    badDMSO_out_top_table <- limma.trend(data = subset, treatment_samples = treatment_samples, control_samples = control_samples)
  }
 
  # plot(plot_volcano(top_table, max.overlaps = 5))
  badDMSO_out_degs <- badDMSO_out_top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull()
  length(intersect(badDMSO_out_degs, drugseq_deg))
  
  badDMSO_out_top_table <- badDMSO_out_top_table %>%
    arrange(p_value_adj, desc(log2FC)) %>%
    mutate(gene = factor(gene, levels = unique(gene)))
  # add a column if there are in the intersect(alldmso_degs, drugseq_deg)
  badDMSO_out_top_table <- badDMSO_out_top_table %>%
    mutate(in_drugseq_deg = ifelse(gene %in% intersect(badDMSO_out_degs, drugseq_deg), "yes", "no"))
  
  plt_ecdf <- ggplot(badDMSO_out_top_table, aes(x = p_value_adj, color = in_drugseq_deg)) +
    stat_ecdf(size = 1) +
    scale_x_continuous(trans = "log10", breaks = c(1e-6,1e-4,5e-2)) +
    labs(x = "Adjusted p-value (log10 scale)", y = "ECDF",
         color = "In DRUGseq DEGs") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    theme_classic()
  
  
  ks_test_results <- ks.test(badDMSO_out_top_table$p_value_adj[badDMSO_out_top_table$in_drugseq_deg=="yes"],
          badDMSO_out_top_table$p_value_adj[badDMSO_out_top_table$in_drugseq_deg=="no"], alternative = "greater")
  
  # ggplot(top_table, aes(x = log2FC, y = -log10(p_value_adj), color = in_drugseq_deg)) +
  #   geom_point(alpha = 0.6, size = 1.2) +
  #   geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  #   geom_vline(xintercept = c(-1, 1), linetype = "dashed") +
  #   scale_color_manual(values = c("no"="#bdbdbd","yes"="#2b8cbe")) +
  #   labs(x = "log2FC", y = expression(-log[10]("adj p")), color = "In DRUGseq DEGs") +
  #   theme_classic()
  # 
  # label a few top overlapping genes
  lab_genes <- badDMSO_out_top_table[badDMSO_out_top_table$in_drugseq_deg=="yes", ] |>
    dplyr::arrange(p_value_adj, dplyr::desc(log2FC)) 
  
  volcano_overlap <- ggplot(badDMSO_out_top_table, aes(x = log2FC, y = -log10(p_value_adj), color = in_drugseq_deg)) +
    geom_point(alpha = 0.6, size = 1.2) +
    geom_text_repel(data = lab_genes, aes(label = gene), size = 3, max.overlaps = 10) +
    scale_color_manual(values = c("no"="#bdbdbd","yes"="#2b8cbe"))+
    theme_classic()
  
  ks <- c(25,50,100,200,500,1000)
  prec_at_k <- sapply(ks, function(k)
    mean(badDMSO_out_top_table$in_drugseq_deg[1:k] == "yes")
  )
  # plot(ks, prec_at_k, type = "b", xlab = "k (top-k by macpie)",
  #      ylab = "Precision@k (fraction in DRUGseq set)")
  
  ks_plot <- ggplot(data.frame(k=ks, prec=prec_at_k), aes(x=k, y=prec)) +
    geom_point() + geom_line() +
    labs(x = "k (top-k by macpie)", y = "Precision@k (fraction in DRUGseq set)")+
    theme_classic()
  
  # alluvial plot
  macpie_tbl <- badDMSO_out_top_table %>% filter(p_value_adj < 0.05 & log2FC>0)
  drugseq_tbl <- ff86_res_toptable %>% filter(p_value_adj < 0.05 & log2FC>0)
  alluvial_plot <- plot_alluvial(macpie_tbl = macpie_tbl, drugseq_tbl = drugseq_tbl)
  
  #return 
  result_list <- list(top_table = badDMSO_out_top_table,
                      num_degs_macpie = length(badDMSO_out_degs),
                      n_overlap = length(intersect(badDMSO_out_degs, drugseq_deg)),
                      ecdf_plot = plt_ecdf,
                      ks_test_results = ks_test_results,
                      volcano_plot = volcano_overlap,
                      ks_plot = ks_plot,
                      alluvial_plot = alluvial_plot)
  return(result_list)
  
})

names(five_dmso_methods_res) <- methods

```

### Summary table

```{r}
#get a table to show number of DEGs and number of overlapping genes with DRUGseq for each method
deg_summary <- map_df(five_dmso_methods_res, function(x) {
  data.frame(
    num_degs_macpie = x$num_degs_macpie,
    n_overlap = x$n_overlap,
    num_degs_DRUGseq = length(drugseq_deg)
  )
}, .id = paste0("macpie_methods"))

deg_summary
```




### ECDF plot

```{r}
#append top_table from each method into a single data frame
good_dmso_top_table <- map_df(five_dmso_methods_res, "top_table", .id = "method")
head(good_dmso_top_table)

#if in_drugseq_deg is "no", change method to "DRUGseq"
good_dmso_top_table<- good_dmso_top_table %>%
  mutate(method = ifelse(in_drugseq_deg == "no", "not_in_DRUGseq", method))

#rename methods
good_dmso_top_table$method <- recode(good_dmso_top_table$method,
                                        "limma_voom" = "macpie:limma_voom",
                                        "DESeq2" = "macpie:DESeq2",
                                        "edgeR" = "macpie:edgeR",
                                        "limma_trend" = "macpie:limma_trend")


ggplot(good_dmso_top_table, aes(x = p_value_adj, color = method)) +
    stat_ecdf(size = 1) +
    scale_x_continuous(trans = "log10", breaks = c(1e-6,1e-4,5e-2)) +
    labs(x = "Adjusted p-value (log10 scale)", y = "ECDF",
         color = "In DRUGseq DEGs") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    theme_classic()


```


<!-- ```{r} -->
<!-- five_dmso_methods_res$limma_voom$ecdf_plot -->
<!-- five_dmso_methods_res$DESeq2$ecdf_plot -->
<!-- five_dmso_methods_res$edgeR$ecdf_plot -->
<!-- five_dmso_methods_res$limma_trend$ecdf_plot -->
<!-- ``` -->
### Overlapping volcano plot

```{r}
five_dmso_methods_res$limma_voom$volcano_plot
five_dmso_methods_res$DESeq2$volcano_plot
five_dmso_methods_res$edgeR$volcano_plot
five_dmso_methods_res$limma_trend$volcano_plot
```




<!-- ### Precision@k -->

<!-- ```{r} -->
<!-- five_dmso_methods_res$limma_voom$ks_plot -->
<!-- five_dmso_methods_res$DESeq2$ks_plot -->
<!-- five_dmso_methods_res$edgeR$ks_plot -->
<!-- five_dmso_methods_res$limma_trend$ks_plot -->
<!-- ``` -->

### Alluvial plot

```{r}
five_dmso_methods_res$limma_voom$alluvial_plot
five_dmso_methods_res$DESeq2$alluvial_plot
five_dmso_methods_res$edgeR$alluvial_plot
five_dmso_methods_res$limma_trend$alluvial_plot
```

<!-- #### Common genes from three methods in macpie -->

<!-- ```{r} -->
<!-- three_methods_degs <- list( -->
<!--   limma_voom = five_dmso_methods_res$limma_voom$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!--   DESeq2 = five_dmso_methods_res$DESeq2$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!--   edgeR = five_dmso_methods_res$edgeR$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!--   limma_trend = five_dmso_methods_res$limma_trend$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull() -->
<!-- ) -->


<!-- # ggvenn( -->
<!-- #   three_methods_degs, -->
<!-- #   fill_color = macpie_colours$discrete,   # optional custom colors -->
<!-- #   stroke_size = 0.5,    # border thickness -->
<!-- #   set_name_size = 4,    # label size -->
<!-- #   text_size = 3         # count label size -->
<!-- # ) +  -->
<!-- #   ggtitle("Genes overlap among macpie methods") -->

<!-- fit <- euler(three_methods_degs) -->
<!-- plot(fit, quantities = TRUE, legend = TRUE)   -->

<!-- ``` -->

<!-- #### Common genes from three methods in macpie and DRUGseq -->

<!-- ```{r} -->
<!-- three_methods_degs <- list( -->
<!--   limma_voom = five_dmso_methods_res$limma_voom$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!--   DESeq2 = five_dmso_methods_res$DESeq2$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!--   edgeR = five_dmso_methods_res$edgeR$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!--   limma_trend = five_dmso_methods_res$limma_trend$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!--   DRUGseq = drugseq_deg -->
<!-- ) -->


<!-- fit <- euler(three_methods_degs) -->
<!-- plot(fit, quantities = TRUE, legend = TRUE)   -->
<!-- ``` -->



## Good DMSO - DRUGseq 

From public DRUG-seq analysis pipeline, authors identified two reference controls: all DMSO wells and the ‘good DMSO’ wells. 

We know these good DMSO wells for batch 24 from their published data:

  - VH02012942: I23, M23
  
  - VH02012944: D23, H23
  
  - VH02012956: I23, J23
  

```{r}
batch24 <- readRDS(paste0(dir,"DRUGseqData/Exp_batch24.Rds"))
names(batch24)
```  

```{r metdata_for_batch24}
#make a combined metadata for three plates
batch24_metadata <- batch24 %>% 
  map_dfr(~ {
    .x$Annotation %>%
      mutate(
        Plate_ID        = plate_barcode,
        Well_ID         = well_id,
        Barcode         = paste0(plate_barcode, "_", well_index),
        Row             = LETTERS[row],
        Column          = as.integer(col),
        Species         = "human",
        Cell_type       = "U2OS",
        Model_type      = "2D_adherent",
        Time            = as.factor(hours_post_treatment),
        Unit            = "h",
        Treatment_1     = cmpd_sample_id,
        Concentration_1 = as.numeric(concentration),
        Unit_1          = unit,
        Sample_type     = if_else(well_type == "SA" & col == 24,
                                  "Positive Control",
                                  well_type)
      )
  })


batch24_metadata <- batch24_metadata%>%select(-c(batch_id, plate_barcode,plate_index, well_id,
                                                 well_index, col, row, biosample_id, external_biosample_id,
                                                 cmpd_sample_id, well_type, cell_line_name, cell_line_ncn, concentration, unit, hours_post_treatment, Sample))
```



```{r count_matrix_for_batch24}
# create a combined UMI matrix for 3 plates
batch24_counts <- batch24 %>%
  map(~ {
    .x$UMI.counts %>%
      as.data.frame() %>% 
      rownames_to_column("gene_id") %>%
      separate(col = gene_id, into = c("gene_name", "chrom"), sep = ",") %>%
      mutate(gene_name = make.unique(gene_name)) %>%
      select(-chrom) %>%
      tibble::column_to_rownames(var = "gene_name") %>%
      as.matrix()
  })

binded_counts <- do.call(cbind, batch24_counts)


  
```



```{r load_data_batch24}
as_mac <- CreateSeuratObject(counts = binded_counts, 
  min.cells = 1, 
  min.features = 1)

as_mac<- as_mac%>% inner_join(batch24_metadata, by = c(".cell"="Barcode"))
```


Filtering steps

Include only good DMSO wells as controls

```{r}

as_mac$combined_id <- paste0(as_mac$Treatment_1,"_", as_mac$Concentration_1)

badDMSO <- as_mac@meta.data %>% filter(Treatment_1 == "CB-43-EP73") %>% 
 filter((Plate_ID == "VH02012942" & !(Well_ID %in% c("I23", "M23"))) |
        (Plate_ID == "VH02012944" & !(Well_ID %in% c("D23", "H23")))|
        (Plate_ID == "VH02012956" & !  (Well_ID %in% c("I23", "J23"))))



keep_wells <- setdiff(rownames(as_mac@meta.data), rownames(badDMSO))


mac_badDSMOremoved <- as_mac[,keep_wells]

mac_badDSMOremoved$combined_id <- str_replace_all(mac_badDSMOremoved$combined_id, "-","_")


```


```{r}
min_sample_num <- min(table(mac_badDSMOremoved$combined_id))
mac_badDSMOremoved <- filter_genes_by_expression(mac_badDSMOremoved,
                                           group_by = "combined_id", min_counts =10,
                                           min_samples = min_sample_num)

```


```{r}
mac_badDSMOremoved[["percent.mt"]] <- PercentageFeatureSet(mac_badDSMOremoved, pattern = "^mt-|^MT-")
mac_badDSMOremoved[["percent.ribo"]] <- PercentageFeatureSet(mac_badDSMOremoved, pattern = "^Rp[slp][[:digit:]]|^Rpsa|^RP[SLP][[:digit:]]|^RPSA")

```








```{r}
methods <- c("limma_voom", "DESeq2", "edgeR", "limma_trend")

good_dmso_methods_res <- lapply(methods, function(m){
  message("Processing method: ", m) 
  # m<-"limma_voom"
  treatment_samples <- "FF_86_NH56_10"
  control_samples <- "CB_43_EP73_0"
  subset <- mac_badDSMOremoved%>%filter(combined_id%in%c(treatment_samples,control_samples))

  batch <- subset$orig.ident
  if (m!="limma_trend"){
      # subset <- filter_genes_by_expression(subset,
      #                        group_by = "combined_id", min_counts = 10,
      #                        min_samples = 1)
     badDMSO_out_top_table <- compute_single_de(subset, treatment_samples, control_samples, method =  m, batch = batch)
  } else{
    badDMSO_out_top_table <- limma.trend(data = subset, treatment_samples = treatment_samples, control_samples = control_samples)
  }
 
  # plot(plot_volcano(top_table, max.overlaps = 5))
  badDMSO_out_degs <- badDMSO_out_top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull()
  length(intersect(badDMSO_out_degs, drugseq_deg))
  
  badDMSO_out_top_table <- badDMSO_out_top_table %>%
    arrange(p_value_adj, desc(log2FC)) %>%
    mutate(gene = factor(gene, levels = unique(gene)))
  # add a column if there are in the intersect(alldmso_degs, drugseq_deg)
  badDMSO_out_top_table <- badDMSO_out_top_table %>%
    mutate(in_drugseq_deg = ifelse(gene %in% intersect(badDMSO_out_degs, drugseq_deg), "yes", "no"))
  
  plt_ecdf <- ggplot(badDMSO_out_top_table, aes(x = p_value_adj, color = in_drugseq_deg)) +
    stat_ecdf(size = 1) +
    scale_x_continuous(trans = "log10", breaks = c(1e-6,1e-4,5e-2)) +
    labs(x = "Adjusted p-value (log10 scale)", y = "ECDF",
         color = "In DRUGseq DEGs") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    theme_classic()
  
  
  ks_test_results <- ks.test(badDMSO_out_top_table$p_value_adj[badDMSO_out_top_table$in_drugseq_deg=="yes"],
          badDMSO_out_top_table$p_value_adj[badDMSO_out_top_table$in_drugseq_deg=="no"], alternative = "greater")
  
  # ggplot(top_table, aes(x = log2FC, y = -log10(p_value_adj), color = in_drugseq_deg)) +
  #   geom_point(alpha = 0.6, size = 1.2) +
  #   geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  #   geom_vline(xintercept = c(-1, 1), linetype = "dashed") +
  #   scale_color_manual(values = c("no"="#bdbdbd","yes"="#2b8cbe")) +
  #   labs(x = "log2FC", y = expression(-log[10]("adj p")), color = "In DRUGseq DEGs") +
  #   theme_classic()
  # 
  # label a few top overlapping genes
  lab_genes <- badDMSO_out_top_table[badDMSO_out_top_table$in_drugseq_deg=="yes", ] |>
    dplyr::arrange(p_value_adj, dplyr::desc(log2FC)) 
  
  volcano_overlap <- ggplot(badDMSO_out_top_table, aes(x = log2FC, y = -log10(p_value_adj), color = in_drugseq_deg)) +
    geom_point(alpha = 0.6, size = 1.2) +
    geom_text_repel(data = lab_genes, aes(label = gene), size = 3, max.overlaps = 10) +
    scale_color_manual(values = c("no"="#bdbdbd","yes"="#2b8cbe"))+
    theme_classic()
  
  ks <- c(25,50,100,200,500,1000)
  prec_at_k <- sapply(ks, function(k)
    mean(badDMSO_out_top_table$in_drugseq_deg[1:k] == "yes")
  )
  # plot(ks, prec_at_k, type = "b", xlab = "k (top-k by macpie)",
  #      ylab = "Precision@k (fraction in DRUGseq set)")
  
  ks_plot <- ggplot(data.frame(k=ks, prec=prec_at_k), aes(x=k, y=prec)) +
    geom_point() + geom_line() +
    labs(x = "k (top-k by macpie)", y = "Precision@k (fraction in DRUGseq set)")+
    theme_classic()
  
  # alluvial plot
  macpie_tbl <- badDMSO_out_top_table %>% filter(p_value_adj < 0.05 & log2FC>0)
  drugseq_tbl <- ff86_res_toptable %>% filter(p_value_adj < 0.05 & log2FC>0)
  alluvial_plot <- plot_alluvial(macpie_tbl = macpie_tbl, drugseq_tbl = drugseq_tbl)
  
  #return 
  result_list <- list(top_table = badDMSO_out_top_table,
                      num_degs_macpie = length(badDMSO_out_degs),
                      n_overlap = length(intersect(badDMSO_out_degs, drugseq_deg)),
                      ecdf_plot = plt_ecdf,
                      ks_test_results = ks_test_results,
                      volcano_plot = volcano_overlap,
                      ks_plot = ks_plot,
                      alluvial_plot = alluvial_plot)
  return(result_list)
  
})

names(good_dmso_methods_res) <- methods

```

### Summary table

```{r}
#get a table to show number of DEGs and number of overlapping genes with DRUGseq for each method
deg_summary <- map_df(good_dmso_methods_res, function(x) {
  data.frame(
    num_degs_macpie = x$num_degs_macpie,
    n_overlap = x$n_overlap,
    num_degs_DRUGseq = length(drugseq_deg)
  )
}, .id = paste0("macpie_methods"))

deg_summary
```




### ECDF plot

```{r}
#append top_table from each method into a single data frame
good_dmso_top_table <- map_df(good_dmso_methods_res, "top_table", .id = "method")
head(good_dmso_top_table)

#if in_drugseq_deg is "no", change method to "DRUGseq"
good_dmso_top_table<- good_dmso_top_table %>%
  mutate(method = ifelse(in_drugseq_deg == "no", "not_in_DRUGseq", method))

#rename methods
good_dmso_top_table$method <- recode(good_dmso_top_table$method,
                                        "limma_voom" = "macpie:limma_voom",
                                        "DESeq2" = "macpie:DESeq2",
                                        "edgeR" = "macpie:edgeR",
                                        "limma_trend" = "macpie:limma_trend")


ggplot(good_dmso_top_table, aes(x = p_value_adj, color = method)) +
    stat_ecdf(size = 1) +
    scale_x_continuous(trans = "log10", breaks = c(1e-6,1e-4,5e-2)) +
    labs(x = "Adjusted p-value (log10 scale)", y = "ECDF",
         color = "In DRUGseq DEGs") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    theme_classic()


```


```{r}
good_dmso_methods_res$limma_voom$ecdf_plot
good_dmso_methods_res$DESeq2$ecdf_plot
good_dmso_methods_res$edgeR$ecdf_plot
good_dmso_methods_res$limma_trend$ecdf_plot
```
### Overlapping volcano plot

```{r}
good_dmso_methods_res$limma_voom$volcano_plot
good_dmso_methods_res$DESeq2$volcano_plot
good_dmso_methods_res$edgeR$volcano_plot
good_dmso_methods_res$limma_trend$volcano_plot
```


<!-- ### Precision@k -->

<!-- ```{r} -->
<!-- good_dmso_methods_res$limma_voom$ks_plot -->
<!-- good_dmso_methods_res$DESeq2$ks_plot -->
<!-- good_dmso_methods_res$edgeR$ks_plot -->
<!-- good_dmso_methods_res$limma_trend$ks_plot -->
<!-- ``` -->

### Alluvial plot

```{r}
good_dmso_methods_res$limma_voom$alluvial_plot
good_dmso_methods_res$DESeq2$alluvial_plot
good_dmso_methods_res$edgeR$alluvial_plot
good_dmso_methods_res$limma_trend$alluvial_plot
```



<!-- #### Common genes from methods in macpie -->

<!-- ```{r} -->

<!-- library(eulerr) -->
<!-- three_methods_degs <- list( -->
<!--   limma_voom = good_dmso_methods_res$limma_voom$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!--   DESeq2 = good_dmso_methods_res$DESeq2$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!--   edgeR = good_dmso_methods_res$edgeR$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!--   limma_trend = good_dmso_methods_res$limma_trend$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull() -->
<!-- ) -->


<!-- ggvenn( -->
<!--   three_methods_degs, -->
<!--   fill_color = macpie_colours$discrete,   # optional custom colors -->
<!--   stroke_size = 0.5,    # border thickness -->
<!--   set_name_size = 4,    # label size -->
<!--   text_size = 3         # count label size -->
<!-- ) +  -->
<!--   ggtitle("Genes overlap among three macpie methods") -->

<!-- fit <- euler(three_methods_degs) -->
<!-- plot(fit, quantities = TRUE, legend = TRUE)    -->

<!-- ``` -->


<!-- #### Common genes from three methods in macpie and DRUGseq -->

<!-- # ```{r} -->
<!-- # three_methods_degs <- list( -->
<!-- #   limma_voom = good_dmso_methods_res$limma_voom$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!-- #   DESeq2 = good_dmso_methods_res$DESeq2$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!-- #   edgeR = good_dmso_methods_res$edgeR$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!-- #   limma_trend = good_dmso_methods_res$limma_trend$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(), -->
<!-- #   DRUGseq = drugseq_deg -->
<!-- # ) -->
<!-- #  -->
<!-- #  -->
<!-- # ggvenn( -->
<!-- #   three_methods_degs, -->
<!-- #   fill_color = macpie_colours$discrete,   # optional custom colors -->
<!-- #   stroke_size = 0.5,    # border thickness -->
<!-- #   set_name_size = 4,    # label size -->
<!-- #   text_size = 3         # count label size -->
<!-- # ) + -->
<!-- #   ggtitle("Genes overlap among macpie methods and DRUGseq results") -->
<!-- #  -->
<!-- # fit <- euler(three_methods_degs) -->
<!-- # plot(fit, quantities = TRUE, legend = TRUE)     -->
<!-- # ``` -->


## Comparison

To compare DEGs with different replicate numbers and different methods

```{r}
methods <- c("limma_voom", "DESeq2", "edgeR", "limma_trend")

get_jaccard <- function(deg_set, drugseq_deg){
  intersection <- length(intersect(deg_set, drugseq_deg))
  union <- length(union(deg_set, drugseq_deg))
  jaccard_index <- intersection / union
  return(jaccard_index)
}

jaccard_index <- lapply(methods, function(m){
  # all dmso
  degs <- methods_res[[m]]$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull()
  jaccard_all <- get_jaccard(degs, drugseq_deg)
  # five dmso
  degs <- five_dmso_methods_res[[m]]$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull()
  jaccard_five <- get_jaccard(degs, drugseq_deg)
  # three dmso
  degs <- good_dmso_methods_res[[m]]$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull()
  jaccard_three <- get_jaccard(degs, drugseq_deg)
  jaccard_index <- data.frame(
    method = m,
    jaccard_all = jaccard_all,
    jaccard_five = jaccard_five,
    jaccard_three = jaccard_three
  )
  return(jaccard_index)
})

df <- as.data.frame(do.call(rbind, jaccard_index))
rownames(df) <- df$method
df <- df %>% select(-method)
colnames(df) <- c("All DMSO", "macpie: 15 DMSO", "DRUGseq: 6 DMSO")
pheatmap::pheatmap(df,
         cluster_rows = FALSE,
         cluster_cols = FALSE,
         display_numbers = TRUE,
         main = "Jaccard Index between macpie DEGs and DRUGseq DEGs")

```

```{r}
library(UpSetR)
all_dmso <- list(
  limma_voom = methods_res$limma_voom$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(),
  DESeq2 = methods_res$DESeq2$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(),
  edgeR = methods_res$edgeR$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(),
  limma_trend = methods_res$limma_trend$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(),
  DRUGseq = drugseq_deg
)
upset(fromList(all_dmso), 
      nsets = 5, 
      order.by = "freq",
      main.bar.color = "black",
      sets.bar.color = "gray23",
      text.scale = c(2, 2, 2, 1.5, 2, 1.5))
```

```{r}
five_dmso <- list(
  limma_voom = five_dmso_methods_res$limma_voom$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(),
  DESeq2 = five_dmso_methods_res$DESeq2$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(),
  edgeR = five_dmso_methods_res$edgeR$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(),
  limma_trend = five_dmso_methods_res$limma_trend$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(),
  DRUGseq = drugseq_deg
)
upset(fromList(five_dmso), 
      nsets = 5, 
      order.by = "freq",
      main.bar.color = "black",
      sets.bar.color = "gray23",
      text.scale = c(2, 2, 2, 1.5, 2, 1.5))
```


```{r}
good_dmso <- list(
  limma_voom = good_dmso_methods_res$limma_voom$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(),
  DESeq2 = good_dmso_methods_res$DESeq2$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(),
  edgeR = good_dmso_methods_res$edgeR$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(),
  limma_trend = good_dmso_methods_res$limma_trend$top_table %>% filter(p_value_adj <0.05 & log2FC>0) %>% select(gene) %>% pull(),
  DRUGseq = drugseq_deg
)
upset(fromList(good_dmso), 
      nsets = 5, 
      order.by = "freq",
      main.bar.color = "black",
      sets.bar.color = "gray23",
      text.scale = c(2, 2, 2, 1.5, 2, 1.5))
```

