<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Generate a box plot to show gene expression (CPM) — plot_counts • macpie</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Generate a box plot to show gene expression (CPM) — plot_counts"><meta name="description" content="This is the function to generate a box plot to show CPM levels of DE genes
among selected treatment samples and control samples."><meta property="og:description" content="This is the function to generate a box plot to show CPM levels of DE genes
among selected treatment samples and control samples."><meta property="og:image" content="https://github.com/PMCC-BioinformaticsCore/macpie.git/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">macpie</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html">Home</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/macpie.html">Overview</a></li>
    <li><a class="dropdown-item" href="../articles/quality_control.html">Quality control</a></li>
    <li><a class="dropdown-item" href="../articles/transcriptional_analyses.html">Transcriptional analyses</a></li>
    <li><a class="dropdown-item" href="../articles/high_throughput_screens.html">Compound screening</a></li>
    <li><a class="dropdown-item" href="../articles/cross_platform_compatibility.html">Cross-platform compatibility</a></li>
    <li><a class="dropdown-item" href="../articles/integration_across_modalities.html">Integration across modalities</a></li>
    <li><a class="dropdown-item" href="../articles/macpie_bioc.html">Working with Bioconductor classes</a></li>
    <li><a class="dropdown-item" href="../articles/assessing_zero_inflation.html">Assessing zero-inflation</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="https://github.com/PMCC-BioinformaticsCore/macpie.git"><span class="fa fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Generate a box plot to show gene expression (CPM)</h1>
      <small class="dont-index">Source: <a href="https://github.com/PMCC-BioinformaticsCore/macpie/blob/HEAD/R/plot_counts.R" class="external-link"><code>R/plot_counts.R</code></a></small>
      <div class="d-none name"><code>plot_counts.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This is the function to generate a box plot to show CPM levels of DE genes
among selected treatment samples and control samples.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_counts</span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group_by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  treatment_samples <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  control_samples <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  color_by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normalisation <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  batch <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>A tidyseurat object merged with metadata. Must contain columns
"Well_ID", "Row", "Column".</p></dd>


<dt id="arg-genes">genes<a class="anchor" aria-label="anchor" href="#arg-genes"></a></dt>
<dd><p>Genes to be plotted</p></dd>


<dt id="arg-group-by">group_by<a class="anchor" aria-label="anchor" href="#arg-group-by"></a></dt>
<dd><p>A column that specifies the treatment group in the input data</p></dd>


<dt id="arg-treatment-samples">treatment_samples<a class="anchor" aria-label="anchor" href="#arg-treatment-samples"></a></dt>
<dd><p>Value in the column "combined_id" representing replicates of treatment samples in the data</p></dd>


<dt id="arg-control-samples">control_samples<a class="anchor" aria-label="anchor" href="#arg-control-samples"></a></dt>
<dd><p>Value in the column "combined_id"  representing replicates of control samples in the data</p></dd>


<dt id="arg-color-by">color_by<a class="anchor" aria-label="anchor" href="#arg-color-by"></a></dt>
<dd><p>A column that specifies the group coloring</p></dd>


<dt id="arg-normalisation">normalisation<a class="anchor" aria-label="anchor" href="#arg-normalisation"></a></dt>
<dd><p>One of "raw", "logNorm", "cpm", "clr", "SCT", "DESeq2",
"edgeR", "RUVg", "RUVs", "RUVr", "limma_voom"</p></dd>


<dt id="arg-batch">batch<a class="anchor" aria-label="anchor" href="#arg-batch"></a></dt>
<dd><p>To indicate patch factor</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a ggplot2 object</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">mini_mac</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="va">mini_mac</span><span class="op">@</span><span class="va">tools</span><span class="op">$</span><span class="va">diff_exprs</span><span class="op">$</span><span class="va">Staurosporine_10</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">plot_counts</span><span class="op">(</span><span class="va">mini_mac</span>, genes <span class="op">=</span> <span class="va">genes</span>, group_by <span class="op">=</span> <span class="st">"combined_id"</span>, </span></span>
<span class="r-in"><span>treatment_samples <span class="op">=</span> <span class="st">"Staurosporine_10"</span>, </span></span>
<span class="r-in"><span>control_samples <span class="op">=</span> <span class="st">"DMSO_0"</span>,</span></span>
<span class="r-in"><span>normalisation <span class="op">=</span> <span class="st">"clr"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Normalizing layer: counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Normalizing across cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % ~calculating    |+                                                 | 1 % ~00s            |++                                                | 3 % ~00s            |++                                                | 4 % ~00s            |+++                                               | 5 % ~00s            |++++                                              | 6 % ~00s            |++++                                              | 8 % ~00s            |+++++                                             | 9 % ~00s            |++++++                                            | 10% ~00s            |++++++                                            | 12% ~00s            |+++++++                                           | 13% ~00s            |++++++++                                          | 14% ~00s            |++++++++                                          | 16% ~00s            |+++++++++                                         | 17% ~00s            |++++++++++                                        | 18% ~00s            |++++++++++                                        | 19% ~00s            |+++++++++++                                       | 21% ~00s            |++++++++++++                                      | 22% ~00s            |++++++++++++                                      | 23% ~00s            |+++++++++++++                                     | 25% ~00s            |+++++++++++++                                     | 26% ~00s            |++++++++++++++                                    | 27% ~00s            |+++++++++++++++                                   | 29% ~00s            |+++++++++++++++                                   | 30% ~00s            |++++++++++++++++                                  | 31% ~00s            |+++++++++++++++++                                 | 32% ~00s            |+++++++++++++++++                                 | 34% ~00s            |++++++++++++++++++                                | 35% ~00s            |+++++++++++++++++++                               | 36% ~00s            |+++++++++++++++++++                               | 38% ~00s            |++++++++++++++++++++                              | 39% ~00s            |+++++++++++++++++++++                             | 40% ~00s            |+++++++++++++++++++++                             | 42% ~00s            |++++++++++++++++++++++                            | 43% ~00s            |+++++++++++++++++++++++                           | 44% ~00s            |+++++++++++++++++++++++                           | 45% ~00s            |++++++++++++++++++++++++                          | 47% ~00s            |+++++++++++++++++++++++++                         | 48% ~00s            |+++++++++++++++++++++++++                         | 49% ~00s            |++++++++++++++++++++++++++                        | 51% ~00s            |++++++++++++++++++++++++++                        | 52% ~00s            |+++++++++++++++++++++++++++                       | 53% ~00s            |++++++++++++++++++++++++++++                      | 55% ~00s            |++++++++++++++++++++++++++++                      | 56% ~00s            |+++++++++++++++++++++++++++++                     | 57% ~00s            |++++++++++++++++++++++++++++++                    | 58% ~00s            |++++++++++++++++++++++++++++++                    | 60% ~00s            |+++++++++++++++++++++++++++++++                   | 61% ~00s            |++++++++++++++++++++++++++++++++                  | 62% ~00s            |++++++++++++++++++++++++++++++++                  | 64% ~00s            |+++++++++++++++++++++++++++++++++                 | 65% ~00s            |++++++++++++++++++++++++++++++++++                | 66% ~00s            |++++++++++++++++++++++++++++++++++                | 68% ~00s            |+++++++++++++++++++++++++++++++++++               | 69% ~00s            |++++++++++++++++++++++++++++++++++++              | 70% ~00s            |++++++++++++++++++++++++++++++++++++              | 71% ~00s            |+++++++++++++++++++++++++++++++++++++             | 73% ~00s            |++++++++++++++++++++++++++++++++++++++            | 74% ~00s            |++++++++++++++++++++++++++++++++++++++            | 75% ~00s            |+++++++++++++++++++++++++++++++++++++++           | 77% ~00s            |+++++++++++++++++++++++++++++++++++++++           | 78% ~00s            |++++++++++++++++++++++++++++++++++++++++          | 79% ~00s            |+++++++++++++++++++++++++++++++++++++++++         | 81% ~00s            |+++++++++++++++++++++++++++++++++++++++++         | 82% ~00s            |++++++++++++++++++++++++++++++++++++++++++        | 83% ~00s            |+++++++++++++++++++++++++++++++++++++++++++       | 84% ~00s            |+++++++++++++++++++++++++++++++++++++++++++       | 86% ~00s            |++++++++++++++++++++++++++++++++++++++++++++      | 87% ~00s            |+++++++++++++++++++++++++++++++++++++++++++++     | 88% ~00s            |+++++++++++++++++++++++++++++++++++++++++++++     | 90% ~00s            |++++++++++++++++++++++++++++++++++++++++++++++    | 91% ~00s            |+++++++++++++++++++++++++++++++++++++++++++++++   | 92% ~00s            |+++++++++++++++++++++++++++++++++++++++++++++++   | 94% ~00s            |++++++++++++++++++++++++++++++++++++++++++++++++  | 95% ~00s            |+++++++++++++++++++++++++++++++++++++++++++++++++ | 96% ~00s            |+++++++++++++++++++++++++++++++++++++++++++++++++ | 97% ~00s            |++++++++++++++++++++++++++++++++++++++++++++++++++| 99% ~00s            |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=00s  </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nenad Bartonicek, Xin Liu, Laura Twomey, Michelle Meier, Richard Lupat, Stuart Craig, David Yoannidis, Jason Li, Tim Semple, Kaylene J Simpson, Mark X Li, Susanne Ramm.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

