---
title: "<img src=\"Cropped_VCFG_Logo2019.png\" style=\"float:right;height:77.1429px;width:152px\"/>`r params$client_name`<br>`r params$project_code` Pathway_Analysis"

params:
  client_name: "Kathleen-Pishas"
  project_code: "PMMsq04"
  
output:
  html_document:
    code_folding: hide
    keep_md: true
    theme: flatly
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 2
    df_print: kable

---



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, tidy = TRUE)
```


<style type="text/css">

body{ /* Normal  */
      font-size: 16px;
  }
td {  /* Table  */
  font-size: 14px;
  white-space: nowrap;
}
h1.title {
  font-size: 32px;
  color: Black;
}
h1 { /* Header 1 */
  font-size: 24px;
  color: Black;
}
h2 { /* Header 2 */
    font-size: 18px;
  color: Black;
}
h3 { /* Header 3 */
  font-size: 18px;
  color: Black;
}
h4 { /* Header 4 */
  font-size: 18px;
  color: Black;
}
h5 { /* Normal  */
  font-size: 16px;
  color: Black;
  class = "author"
}
h6 { /* Normal  */
  font-size: 16px;
  color: Black;
  class = "data"
}
a, a:hover {
  color: #7d2b8b
}
code.r{ /* Code block */
    font-size: 12px;
}
pre { /* Code block - determines code spacing between lines */
    font-size: 14px;
}
</style>


<style type="text/css">
#TOC { /* Normal  */
  color: Black; 
  border-color: #7d2b8b
}
a {
color: Black;
}
.list-group-item.active, .list-group-item.active:focus, .list-group-item.active:hover {
    background-color: #7d2b8b;
    border-color: #7d2b8b;
}
</style>

<style>
div.a {
  text-indent: 50px;
}
</style>

<style>
.nav-pills>li>a {
    position: relative;
    display: block;
    color: #3c2372;
    border: 1px solid #7d2b8b;
    background-color: #ffffff;

    }
.nav-pills>li.active>a, .nav-pills>li.active>a:hover, .nav-pills>li.active>a:focus {
    color: #ffffff;
    background-color: #7d2b8b;
    border: 1px solid #ffffff;

    }
</style>
<style>

.tabset-dropdown > .nav-tabs {
	display: inline-block;
	max-height: 500px;
	min-height: 44px;
	overflow-y: auto;
	background: white;
	border: 1px solid #7d2b8b;
	border-radius: 4px;
	color:#7d2b8b;
}
</style>




# Enrichment Analysis

```{r}

suppressMessages(library(dplyr))
suppressMessages(library(httr2))
suppressMessages(library(ggplot2))
suppressMessages(library(DT))
suppressMessages(library(here))
suppressMessages(library(knitr))

#Functions of enrichment analysis
source(here::here("code/PathwayFunctions.R"))

```




## Connect and extract datasets from EnrichR

Here are the datasets available on EnrichR that we can pull down. 

```{r}
url<-enrichr_url("enrichr")
summarydf<-GS_summary(url)
datatable(summarydf, caption="Table of all datasets on Enrichr")
```

## KEGG_2021_Human

For this enrichment analysis, number of background genes is set to the number of genes in human genome (same as in Enrichr).

We set this background gene number quite flexible, you can provide your own number to perform the enrichment test. 

```{r}
kegg<-geneset_download(url,"KEGG_2021_Human")

```


Checking files in the folder of `r params$project_code`: 


```{r}
wrdir<-"/Users/liuxin/Documents/MACseq/"
list_DEGs <- grep("*.csv", list.files(paste0(wrdir, "/", params$client_name, "/DEGresults/",
    params$project_code,"/")), value = TRUE)
list_DEGs

```



### Pathway analysis 

Genes of interest for pathway analysis: are genes with adjusted P values <=0.05 and log 2 fold changes>0 - as these are the ones up-regulated with treatment 



```{r}

list_pathway<-lapply(list_DEGs, function(x){
  namelist<-as.character(x)
  namelist<-substring(namelist, 1,nchar(namelist)-4)
  x<-read.csv(paste0(wrdir, "/", params$client_name, "/DEGresults/",params$project_code,"/",x), header = TRUE)
  #bg<-length(rownames(x))
  x<-x[!is.na(x$FDR),]
  deg<-x[x$FDR<=0.05 & x$logFC>0,]
  kegg_exp<-hyper_enrich_bg(deg$X, kegg, background = "human")
  # assign(namelist, kegg_exp)
  # namelist<-kegg_exp
})

names(list_pathway)<-substring(list_DEGs, 1,nchar(list_DEGs)-4)

```



### Bar plots

Full results of enrichment analyses were generated in CSV files.

Here, we're having a quick look of top 20 enriched pathways (pathways with smallest adjusted p values) for every data set. 

Note: some data sets have no pathway with adjusted p values <= 0.05, also showing in this quick look step. 

Hits refer to number of differentially expressed genes are annotated in the given pathways. 



```{r results='asis'}
lapply(list_pathway, function(x){
  plot_enrich(x, numTerms = 20)

})
```




```{r}
out_dir<-paste0(wrdir,params$client_name,"/Pathway/",params$project_code,"/")
system(paste0("mkdir -p ",out_dir))

lapply(1:length(list_pathway), function(x){
  write.csv(list_pathway[x], paste0(out_dir,names(list_pathway[x]),"_KEGG.csv"))
}) 



```


<br>

<br>

  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.13.0/css/all.css">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.13.0/css/v4-shims.css">

  &nbsp;
<hr />
<p style="text-align: center;">Analysed by Xin Liu</a></p>
<p style="text-align: center;"><span style="color: #7d2b8b;"><em>Xin.Liu@petermac.org</em></span></p>
<p style="text-align: center;"><a href="https://www.petermac.org/research/core-facilities/victorian-centre-functional-genomics"><span style="color: #7d2b8b;">Victorian Centre for Functional Genomics</span></a></p>


  <!-- Needs the "fab" prefix, V5 <i class="fab fa-twitter"></i> -->
<p style="text-align: center;">  
  <a href="https://twitter.com/VCFGconnect"> <i class="fa fa-twitter"></i></a>
</p>

&nbsp;

