---
title: "<img src=\"Cropped_VCFG_Logo2019.png\" style=\"float:right;height:77.1429px;width:152px\"/>`r params$client_name`<br>`r params$project_code`- plate QC: NEW format v1<br>"

params:
  client_name: "Kathleen_Pishas"
  project_code: "PMMsq03"
  
output: 
  html_document:
    keep_md: true
    theme: flatly
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 3
    code_folding: hide
    df_print: kable
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, tidy = TRUE)
```


<style type="text/css">

body{ /* Normal  */
      font-size: 16px;
  }
td {  /* Table  */
  font-size: 14px;
  white-space: nowrap;
}
h1.title {
  font-size: 32px;
  color: Black;
}
h1 { /* Header 1 */
  font-size: 24px;
  color: Black;
}
h2 { /* Header 2 */
    font-size: 18px;
  color: Black;
}
h3 { /* Header 3 */
  font-size: 18px;
  color: Black;
}
h4 { /* Header 4 */
  font-size: 18px;
  color: Black;
}
h5 { /* Normal  */
  font-size: 16px;
  color: Black;
  class = "author"
}
h6 { /* Normal  */
  font-size: 16px;
  color: Black;
  class = "data"
}
a, a:hover {
  color: #7d2b8b
}
code.r{ /* Code block */
    font-size: 12px;
}
pre { /* Code block - determines code spacing between lines */
    font-size: 14px;
}
</style>


<style type="text/css">
#TOC { /* Normal  */
  color: Black; 
  border-color: #7d2b8b
}
a {
color: Black;
}
.list-group-item.active, .list-group-item.active:focus, .list-group-item.active:hover {
    background-color: #7d2b8b;
    border-color: #7d2b8b;
}
</style>

<style>
div.a {
  text-indent: 50px;
}
</style>

<style>
.nav-pills>li>a {
    position: relative;
    display: block;
    color: #3c2372;
    border: 1px solid #7d2b8b;
    background-color: #ffffff;

    }
.nav-pills>li.active>a, .nav-pills>li.active>a:hover, .nav-pills>li.active>a:focus {
    color: #ffffff;
    background-color: #7d2b8b;
    border: 1px solid #ffffff;

    }
</style>
<style>

.tabset-dropdown > .nav-tabs {
	display: inline-block;
	max-height: 500px;
	min-height: 44px;
	overflow-y: auto;
	background: white;
	border: 1px solid #7d2b8b;
	border-radius: 4px;
	color:#7d2b8b;
}
</style>





## Aim 

The aim is to check quality of MACSeq reads, plate layout, consistency among controls and samples, and to explore any hidden correlations among different factors in this experiment. 


***

<br>


## Overview

Heatmap is used to check plate layout. We both check at whole plate level and at individual cell line level to ensure we're able to check and show any inconsistencies in different wells/cell lines. Then principal component analysis ([PCA method](../docs/MACseq_analysis_methods.html#Principal_component_analysis_(PCA):)) is employed to reveal and expose any hidden relationships among different factors, suggesting which ones might be highly correlated and which might be independent. Gene expression and read counts distribution are examined to look for any outliers, or inconsistencies between samples that belong to same treatment.  



***

<br>




## Alignment QC 

This file shows the quality of FASTQ files.

[multiQC results](../docs/PMMsq03_multiqc_report.html)

**Comments**: quality of the FASTQ files are good.  Uniquely mapped reads is 342 millions (82% of the total reads are uniquely aligned).   



***

<br>


```{r}
library(tidyverse)
library(Seurat)
library(SeuratObject)
library(gtools)
library(tidyseurat)
library(reshape2)
library(gridExtra)
library(rlang)
library(knitr)
library(DT)
library(RColorBrewer)
library(platetools)
library(RUVSeq)
source(here::here("code/QCFunctions.R"))

```



```{r}

#from cluster
dir<-"/Volumes/bioinf/team_folders/MGC_VCFG/macseq"
metadata<-read.csv(paste0(dir,"/metadata/",params$client_name,"/",params$project_code,"/","corrected_",params$project_code,".csv"), sep = ",")
#change metadata order
metadata<-metadata[order(metadata$Column),]



#import data
nextflow_dir<-paste0(dir,"/results/",params$project_code)
plateIDs<-unique(metadata$Plate_ID)

#find out which nextflow directory to use: probably the latest with the finished analysis
countmatrix_dirs<-list.files(nextflow_dir,recursive = TRUE,full.names=T,pattern="matrix.mtx.gz")
countmatrix_dirs<-countmatrix_dirs[grep(plateIDs,countmatrix_dirs)]
most_recent_countmatrix_folder <- countmatrix_dirs[which.max(file.info(countmatrix_dirs)$mtime)] %>%
  str_replace_all("/matrix.mtx.gz","")
# #change back to raw
most_recent_countmatrix_folder<-str_replace(most_recent_countmatrix_folder,"filtered","raw")


raw_counts<-Read10X(data.dir=most_recent_countmatrix_folder)
keep <- rowSums(cpm(raw_counts)>10) >= 2
raw_counts <- raw_counts[keep,]
# dim(raw_counts)

#filter for the minimum number of reads per gene per sample, as per edgeR 
mac <- CreateSeuratObject(counts = raw_counts, 
                          project = params$project_code, 
                          min.cells = 1, 
                          min.features = 1) 

#add location metadata
mac<- mac %>%
  inner_join(metadata,by=c(".cell"="Barcode")) 

#exclude edges 
#Row A, P
#Column 1, 24

mac_noedge<-subset(mac, subset=Row!="A" & Row!="P" & Column!="1" & Column!="24")
# dim(mac_noedge)



```



## Plate layout 

### Annotation 

This is the annotation of the plate layout. 


![](../docs/PMC220_PlateLayout.png){width=2000, height=200}


**Drug Reference**

```{r, results='asis'}
drug_ref<-data.frame(Drug_plate_ID=c(1:21),
           Drug=c("Pelitinib","Dacomitinib","Neratinib","Avitinib (maleate)","AEE788","Cobimetinib","Mirdametinib",
              "Trametinib","TAK-733","Selumetinib","RO4987655","Pimasertib","Refametinib","Binimetinib","Avutometinib",
              "Carboplatin","Paclitaxel","Tucidinostat","Palbociclib","Dabrafenib","Defactinib"),
           Class=c("EGFR; Src",rep("EGFR",4), "Apoptosis; MEK","Apoptosis; Autophagy; MEK","Apoptosis; Autophagy; MEK",
                   "MAPK/ERK Pathway","Apoptosis; MAPK/ERK Pathway", rep("MAPK/ERK Pathway",3),"Autophagy; MAPK/ERK Pathway",
                   "MAPK/ERK Pathway","Autophagy; DNA Alkylator/Crosslinker;","Microtubule","HDAC","CDK","Raf","FAK"))

datatable(drug_ref, caption = "Drug Reference")

```




***

<br>

### Heatmap {.tabset .tabset-fade .tabset-pills}

Number of RNA read counts and number of expressed genes are plotted. 

We expect to see similar colour in the replicate wells, suggesting replicate wells have similar number of RNA read counts or number of expressed genes. For control wells, we expect to see higher RNA read counts in DMSO control wells than in positive controls such as Staurosporine 1uM.


#### Whole plate


```{r, fig.height=4, fig.width=11}
p1<-plot_grid_platelayout(mac@meta.data, "nCount_RNA",title_label = "whole plate",legend_label="Number of reads")
p2<-plot_grid_platelayout(mac@meta.data, "nFeature_RNA",title_label = "whole plate",legend_label="Number of genes")


grid.arrange(p1,p2,ncol=2)
```


**Comments: **

**Whole plate**: The biggest differences of read counts and gene counts are by three different cell lines. The first cell line (VOA-4698) has the highest read counts and cell counts. The middle cell line (VOA-10841) has the lowest numbers of both RNA reads and gene counts. Higher read counts along the edges is due to edge effect, which means the wells on the edges take up more reads.  


```{r,include=FALSE}
comment_generator<-function(x){
  sapply(x, function(y){
    ifelse(y==cell_types[1],
           paste0("**Comments:** wells along the edges show strong edge effect, as most of these wells show higher RNA read counts than inner wells. For DMSO control wells (B08, C08) show consistent number of RNA read counts. For Staurosporine, we have consistent RNA read counts between replicates: Staurosporine 1uM wells show low RNA counts than 0.1uM wells. For Mitomycin (H08 to K08), RNA read counts are quite variable. For Doxorubicin (L08 to O08), Doxorubicin 1uM show much lower RNA read counts than Doxorubicin 0.1uM.\\\n Wells along the edges show relatively higher gene counts. Staurosporine 1uM wells have the lowest gene counts, and Staurosporine 0.1uM wells have higher gene counts. Mitomycin don't show any differences in the gene counts. Only one well of Doxorubicin 1uM shows lower number of genes than the other replicates.\\\n"),
           ifelse(y==cell_types[2],
           paste0("**Comments:** wells on the top and bottom rows have higher RNA read counts, which is due to edge effect. For DMSO control wells (B15, C15) show consistent number of RNA read counts. Staurosporine at 1 uM and 0.1uM (D15 to G15) show similar level of RNA read counts between replicates (lower counts in 1uM than in 0.1uM). For Mitomycin (H15 to K15), RNA read counts are quite similar. For Doxorubicin (L15 to O15), Doxorubicin 1uM wells (variation between replicates) show lower RNA read counts than Doxorubicin 0.1uM.\\\n Wells allong the edges show relatively higher gene counts. Staurosporine 1uM wells have the lowest gene counts, and Staurosporine 0.1uM wells have higher gene counts. Mitomycin wells don't show any differences in the gene counts. Only one well of Doxorubicin 1uM shows lower number of genes than the other replicates.\\\n"),
           paste0("**Comments:** wells on the top and bottom rows have higher RNA read counts, which is due to edge effect. For DMSO control wells (B22, C22) show consistent number of RNA read counts. Staurosporine at 1 uM and 0.1uM (D22 to G22) show lower level of RNA read counts at 1uM wells. For Mitomycin (H22 to K22), RNA read counts are quite close. For Doxorubicin (L22 to O22), Doxorubicin 1uM wells show lower RNA read counts than Doxorubicin 0.1uM.\\\n Wells allong the edges show relatively higher gene counts. Staurosporine 1uM wells have the lowest gene counts, and Staurosporine 0.1uM wells have higher gene counts. Both Mitomycin and Doxorubicin don't show much differences in the gene counts.
 \\\n")))
      
    
  })
}
```



```{r, results='asis', fig.height=4, fig.width=11}
cell_types<-unique(mac@meta.data$Cell_type)

for (cline in cell_types){
  cat("\n")
  cat(paste0("#### ", cline,"\n"))
  cat("\n")
  p1<-plot_grid_platelayout(mac@meta.data%>%filter(Cell_type==cline), "nCount_RNA",title_label = cline,legend_label="Number of reads")
  p2<-plot_grid_platelayout(mac@meta.data%>%filter(Cell_type==cline), "nFeature_RNA",title_label = cline,legend_label="Number of genes")

  grid.arrange(p1,p2,ncol=2)
  cat("\n")
  cat(comment_generator(cline),"\n")
  cat("\n\n")
}


```


### {-}



Cell number plot can't be plotted, as cell number isn't available for this plate. 


***

<br>




## Principal component analysis {.tabset .tabset-fade .tabset-pills}


Principal component analysis (PCA) is used to to explore and examine any correlations between features. By identifying groups of features that contribute most to the overall data variance, PCA can reveal underlying structures and potential redundancies within your data. This can expose hidden relationships between features, suggesting which ones might be highly correlated (measuring similar aspects) and which might be independent.


### Whole plate


```{r, fig.width=14, fig.height=8}
mac<-mac%>%mutate(edge_wells=case_when((Column ==1 | Column == 24|Row=="A" | Row=="P") ~ "TRUE",
                                      TRUE ~ "FALSE")) 
                                      

mac<-clustering(mac)
plot_clusters(mac)

```

**Comments**: There are 3 big clusters, which can be explained by 3 cell lines (in the Cell_type plot). Within each cell type, there are two sub-clusters (one cell line cluster has 3 sub-clusters in seurat_clusters), which can be partially explained by edge wells in each cell line. In order to get a better resolution to compare transcriptional differences among different drugs in each cell line, we need to split into each cell line, and exclude wells along the edges to examine and compare. 

Other than cell lines and edge wells, there aren't any strong correlations between mitochondrial percentage (percent.mito), ribosomal percentage (percent.ribo), rows (Row), or columns (Columns) with the clustering. This is good, which suggesting none of these features contribute to the clustering. 




```{r}
plot_clusters_cline <- function(mac) {
  p1 <- FeaturePlot(mac, features = "nCount_RNA")
  p2 <- FeaturePlot(mac, features = "percent.mt")
  p3 <- FeaturePlot(mac, features = "percent.ribo")
  p4 <- DimPlot(mac, group.by = "Sample_type")
  p4 <- DimPlot(mac, group.by = "Row",cols = c(brewer.pal(8,"Dark2"),brewer.pal(8, "Set1")))
  p5 <- DimPlot(mac, group.by = "Column", label = TRUE, repel = TRUE, cols = c(brewer.pal(8,"Dark2"),brewer.pal(8, "Set1")))
  p6 <- DimPlot(mac, group.by = "Concentration_1")
  p7 <- DimPlot(mac, group.by = "seurat_clusters")
  # p8 <- DimPlot(mac, group.by = "Cell_type")
  p8 <- DimPlot(mac, group.by = "Sample_type", cols=c("orange", "grey80","black"), pt.size = 0.5)
  # p9 <-DimPlot(mac_cline, group.by = "Treatment_1", cols = c(brewer.pal(8,"Dark2"),brewer.pal(8, "Set1"), brewer.pal(12,"Paired")))+guides(color = guide_legend(override.aes = list(size=1), ncol=1, cex=1))+theme(FontSize(-1))
  plot(CombinePlots(plots = list(p1, p2, p3, p4, p5, p6, p7, p8), nrow = 3))
}

comment_generator<-function(x){
  sapply(x, function(y){
    ifelse(y==cell_types[1],
           paste0("**Comments:** Four clusters are showed in seurat_clusters plot. Separation on PC1, could be due to ribosomal percentage. Vehicle_control DMSO wells are not clustered closely together - they're separated far apart (black dots in Sample_type plot).\\\n Apart from ribosomal percentaage, other factors don't seem to be correlated with the clustering, which are good. \\\n"),
           ifelse(y==cell_types[2],
           paste0("**Comments:** Three clusters are in seurat_clusters plot. Clusters 1 and 2 that consist of columns 11 & 12 are well separated from other cluster on PC1. The rest of columns are grouped into cluster 0. Separation of wells on PC2 could be due to ribosomal percentage. We do see clustering by columns for this cell line. We expect to see wells with same drug and same concentration that are clusters together, but we don't necessarily like to see wells clustered together showing any correlation with columns (unless the plate layout is designed that way), especially both positive and negative controls in column 15 that are grouped together.\\\n Vehicle_control DMSO wells are clustered together here. \\\n"),
           paste0("**Comments:** Four clusters in seurat_clusters plot. Columns 16 to 19 are clustered in clusters 0 & 2, and from columns 20 to 23 are in cluster 1 & 3. Again, we see clustering by columns for the cell line. Separation of wells on PC2 seems to be due to ribosomal percentage.\\\n DMSO and all the wells of media in column 23 are vehicle_control in Sample_type plot, and these only in clusters 1 & 3. *Column 23 is media.* \\\n")))
      
    
  })
}



```





```{r,fig.width=14, fig.height=8, results='asis'}

clustering_cell_types<-lapply(cell_types, function (x){
  cat("\n")
  cat(paste0("### ", x,"\n"))
  cat("\n")
  mac_cline<-mac_noedge%>%subset(Cell_type==paste0(x))
  mac_cline<-clustering(mac_cline)
  
  cat("\n")
  plot_clusters_cline(mac_cline)
  cat(comment_generator(x),"\n")
  
  cat("\n\n")
})

```



## {-}


***

<br>


## Sample type distribution 

This plot is to show distribution of RNA read counts in samples, positive and vehicle controls. We expect to see positive controls with high concentration have lower RNA read counts than those with low concentration. Vehicle controls show relatively high RNA read counts. 


```{r}


mac_noedge<-mac_noedge%>%mutate(Conc_colour=case_when((Concentration_1==0)~"black",
                                                      (Concentration_1==0.1)~"grey",
                                                      TRUE ~ "red"
  
))
mac_noedge$Cell_type <- factor(mac_noedge$Cell_type, levels = c("VOA-4698", "VOA-10841",
    "AOCS-2"))
mac_noedge %>%
    subset(., subset = Column != "23") %>%
    ggplot(aes(reorder(Sample_type, nCount_RNA, FUN = median), nCount_RNA)) + geom_violin() +
    facet_wrap(vars(Cell_type, nrow = 1)) + geom_jitter(width = 0.1, alpha = 1, aes(color=Concentration_1)) +theme_classic()+
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + xlab("") +scale_colour_gradient(low="grey60",high="purple") 
    
```


**Comments:** 

**VOA-4698**: Two DMSO samples in vehicle controls show a small difference in RNA reads (compared to same cell line in PMMSq04 plate). In the positive controls, samples with high concentration tend to have lower RNA read counts as expected. 

**VOA-10841**: Two DMSO samples (vehicle controls) are tightly together, which expected. Positive control samples with high concentration tend to have lower RNA read counts as expected. 

**AOCS-2**: Two DMSO samples (vehicle controls) are tighly together, which as expected. In the positive controls, samples with high concentration tend to have lower RNA read counts as expected. 


***

<br>


## Gene expression {.tabset .tabset-fade .tabset-pills}


The Relative Log Expression (RLE) plot is a useful diagnostic plot to visualize the differences between the distributions of read counts across samples [RLE method](../docs/MACseq_analysis_methods.html#Relative_log_expression_(RLE):). 

It shows the boxplots of the log-ratios of the gene-level read counts of each sample to those of a reference sample (defined as the median across the samples). Ideally, the distributions should be centered around the zero line and as tight as possible. Clear deviations indicate the need for normalization and/or the presence of outlying samples.

For each RLE plot, replicates from the same condition are plotted next to each other. We expect to see these replicates showing very similar distributions/ranges in their boxes. 

Box plot of number counts is another useful diagnostic way to visualize the difference between the RNA read counts across replicates. We would like to see small variation of the RNA read counts in replicates from the same group. 

Table of read counts is also available to highlight any treatment/control has greater variation in the given cell line. 


### DMSO 


```{r}
DMSO_counts <- mac_noedge %>%
    filter(Treatment_1 == "DMSO")

RLEplot(DMSO_counts@assays$RNA$counts, DMSO_counts$Well_ID, feature = DMSO_counts$Cell_type)
```

**Comments:** From the RLE plot, box plots are expected to have a small interquartile range and their median values are centering around 0 line. Differences of DMSO expression correlate with the differences of RNA read counts among all three cell lines.

**VOA-4698:** gene expression levels are higher in C08 than B08.

**VOA-10841:** gene expression levels are quite consistent in B15 and C15. 

**AOCS-2:** the variation between B22 and C22 is small. 


```{r, include=FALSE}
datatable(cars)
comment_generator<-function(x){
  sapply(x, function(y){
    ifelse(y==cell_types[1],
           paste0("**Comments:** For controls, Mito 0.1uM and Dox 1uM have greater standard deviation than the rest of wells. For treatments, 8 different treatment_conc conditions also highlighted out.\\\n RLE plot reflects those conditions that are highlighted out, such as Dox 1uM, Mito 0.1uM, Neratinib 1uM, RO4987655 1uM."),
           ifelse(y==cell_types[2],
           paste0("**Comments:** For controls, Dox 0.1uM and Dox 1uM have greater standard deviation than the rest of wells. For treatments, 8 different treatment_conc conditions also highlighted out.\\\n RLE plot reflects those conditions that are highlighted out, such as Dox 1uM, Avitinib (maleate) 1uM"),
           paste0("**Comments:** No controls are flagged out (above the cell line sd), which are good.  For treatments, 11 different treatment_conc conditions also highlighted out, such as Pimasertib 0.1uM, Refametinib 0.1uM. \\\n")))
      
    
  })
}


```


```{r, results='asis', fig.width=16, fig.width=8}
mac_noedge_sub<-mac_noedge %>%subset(., subset = Column != "23")
clustering_cell_types<-lapply(cell_types, function (x){
  cat("\n")
  cat(paste0("### ", x,"\n"))
  cat("\n")
  
  mac_cline<-mac_noedge_sub%>%subset(Cell_type==paste0(x))
  meta<-mac_cline@meta.data
  meta$Treatment_conc<-paste0(meta$Treatment_1,".",meta$Concentration_1)
  meta<-meta%>%group_by(Treatment_conc)%>%mutate(Replicate=1:n())
  meta$Sample_ID<-paste0(meta$Treatment_conc,"_Rep",meta$Replicate)
  meta$Replicate <- paste0("Replicate", meta$Replicate)
  reads_sd<-sd(mac_cline$nCount_RNA)
  cat("\n")
  cat(paste0("#### ", "sd of all samples: ",reads_sd), "\n")
  cat("\n")
  #controls
  meta_contols<-meta%>%filter(Sample_type!="Sample")
  
  nRNA <- meta_contols %>%
    select(nCount_RNA, Treatment_conc, Replicate)
  nRNA_long <- dcast(nRNA, Treatment_conc ~ Replicate, value.var = "nCount_RNA")
  nRNA_long <- nRNA_long %>%
    group_by(Treatment_conc) %>%
    mutate(sd = sd(c(Replicate1, Replicate2)))%>%
    mutate(cutoff_reads=case_when((sd>reads_sd)~"1",
                                  TRUE~ "0"))
  
  
  plot(ggplot(meta, aes(Treatment_conc, nCount_RNA)) + geom_boxplot(outlier.colour = "NA") +
    geom_jitter(width = 0.1, aes(color = Treatment_conc)) + theme_bw() + theme(axis.text.x = element_text(angle = 90,
    vjust = 0.5, hjust = 1), legend.position = "none"))
  cat("\n\n")
  
  cat("*Standard deviation of all samples in the cell line is calculated, and then is used to indicate if or not the standard deviation (sd) for each treatment_conc is above the sd from the entire cell line. If sd of a given treatment_conc is greater than the sd of the entire cell line, it is marked as 1 (otherwise as 0) in the cutoff_reads column and its corresponding treatment_conc is coloured in yellow.* ")
  
  cat("\n")
  cat("*As we only have two data points in each treatment_conc group, standard deviation may not provide comprehensive insights. It's just a way to flag out those have great variation.*")
  
  cat("\n")
  
  print(htmltools::tagList(datatable(nRNA_long, caption = "Controls")%>%
                             formatStyle('Treatment_conc','cutoff_reads', backgroundColor = styleEqual(c(0,1), c("white","yellow")))))
  cat("\n")
  
  #samples
  meta_samples<-meta%>%subset(Sample_type=="Sample")
  
  nRNA <- meta_samples %>%
    select(nCount_RNA, Treatment_conc, Replicate)
  nRNA_long <- dcast(nRNA, Treatment_conc ~ Replicate, value.var = "nCount_RNA")
  nRNA_long <- nRNA_long %>%
    group_by(Treatment_conc) %>%
    mutate(sd = sd(c(Replicate1, Replicate2)))%>%
        mutate(cutoff_reads=case_when((sd>reads_sd)~"1",
                                  TRUE~ "0"))
  
    print(htmltools::tagList(datatable(nRNA_long, caption = "Samples")%>%
                             formatStyle('Treatment_conc','cutoff_reads', backgroundColor = styleEqual(c(0,1), c("white","yellow")))))

  cat("\n")

    
  
  plot(RLEplot(mac_cline@assays$RNA$counts, meta$Sample_ID, feature =meta$Sample_ID)+ 
         scale_fill_manual(values = colorRampPalette(brewer.pal(12, "Paired"))(98))+ theme(legend.position="none",axis.text.x = element_text(size = 8)))
  
  cat(comment_generator(x),"\n")
  cat("\n\n")
})


```



## {-}

***

<br>


## Key outcomes

- Higher RNA read counts on the top and bottom of the entire plate could be due to edge effects. Edge effect still exists when splitting into individual cell line. 

- **Dox** 1uM and **Mito** at 0.1uM in VOA-4698 have greater standard deviations than most of the samples. 

- **Dox** 0.1uM and 1uM in VOA-10841 show big standard deviations than most of the samples.

- In all three cell lines, variations of **DMSO** and **Stauro** are within reasonable range.

- From the PCA results, clustering of samples do correlate with particular columns. It could be due to same type of treatments in these columns, which causing them cluster together. Or it could be due to column effect. 





