<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>differential_expression • macpie</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="differential_expression">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">macpie</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html">Home</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/metadata_check.html">Metadata check</a></li>
    <li><a class="dropdown-item" href="../articles/quality_control.html">Quality control</a></li>
    <li><a class="dropdown-item" href="../articles/differential_expression.html">Transcriptional analyses</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Compound screening</h6></li>
    <li><a class="dropdown-item" href="../articles/macpie_conversion.html">Interoperability</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="https://github.com/PMCC-BioinformaticsCore/macpie.git"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>differential_expression</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/PMCC-BioinformaticsCore/macpie.git/blob/HEAD/vignettes/differential_expression.Rmd"><code>vignettes/differential_expression.Rmd</code></a></small>
      <div class="d-none name"><code>differential_expression.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PMCC-BioinformaticsCore/macpie.git">macpie</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">enrichR</span><span class="op">)</span></span>
<span><span class="co">#&gt; Welcome to enrichR</span></span>
<span><span class="co">#&gt; Checking connections ...</span></span>
<span><span class="co">#&gt; Enrichr ... Connection is Live!</span></span>
<span><span class="co">#&gt; FlyEnrichr ... Connection is Live!</span></span>
<span><span class="co">#&gt; WormEnrichr ... Connection is Live!</span></span>
<span><span class="co">#&gt; YeastEnrichr ... Connection is Live!</span></span>
<span><span class="co">#&gt; FishEnrichr ... Connection is Live!</span></span>
<span><span class="co">#&gt; OxEnrichr ... Connection is Live!</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">parallel</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/myles-lewis/mcprogress" class="external-link">mcprogress</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">PoiClaClu</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SeuratObject</span></span>
<span><span class="co">#&gt; Loading required package: sp</span></span>
<span><span class="co">#&gt; 'SeuratObject' was built under R 4.3.0 but the current version is</span></span>
<span><span class="co">#&gt; 4.3.3; it is recomended that you reinstall 'SeuratObject' as the ABI</span></span>
<span><span class="co">#&gt; for R may have changed</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SeuratObject'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, t</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/tidyseurat" class="external-link">tidyseurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: ttservice</span></span>
<span><span class="co">#&gt; ========================================</span></span>
<span><span class="co">#&gt; tidyseurat version 0.8.0</span></span>
<span><span class="co">#&gt; If you use TIDYSEURAT in published research, please cite:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Mangiola et al. Interfacing Seurat with the R tidy universe. Bioinformatics 2021.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; This message can be suppressed by:</span></span>
<span><span class="co">#&gt;   suppressPackageStartupMessages(library(tidyseurat))</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt; To restore the Seurat default display use options("restore_Seurat_show" = TRUE) </span></span>
<span><span class="co">#&gt; ========================================</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'tidyseurat'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following object is masked from 'package:ttservice':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     plot_ly</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p><strong>Data access</strong><br>
The full dataset (&gt;10 MB of <code>.rds</code> files) is currently
under restricted release and will become publicly available upon
publication (at Zenodo).<br>
In the meantime, please contact us for early access.</p>
</blockquote>
<div class="section level2">
<h2 id="differential-gene-expression">Differential gene expression<a class="anchor" aria-label="anchor" href="#differential-gene-expression"></a>
</h2>
<p><strong>Key points</strong>:</p>
<ul>
<li>use <strong>compute_single_de</strong> to perform a differential
expression analysis for one treatment group vs control</li>
<li>use <strong>compute_multi_de</strong> to perform differential
expression analyses for all treatment groups vs control</li>
<li>use volcano plot, box plot and heatmap to show results from the
analyses and visualise gene expression levels<br>
</li>
<li>use <strong>enrichr</strong> for pathway enrichment analysis</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">project_name</span> <span class="op">&lt;-</span> <span class="st">"PMMSq033"</span></span>
<span><span class="va">project_metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/PMMSq033_metadata.csv"</span>, package <span class="op">=</span> <span class="st">"macpie"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load metadata</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_metadata.html">read_metadata</a></span><span class="op">(</span><span class="va">project_metadata</span><span class="op">)</span></span>
<span><span class="va">metadata</span><span class="op">$</span><span class="va">Time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">Time</span><span class="op">)</span></span>
<span><span class="va">metadata</span><span class="op">$</span><span class="va">Concentration_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">Concentration_1</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Import raw data</span></span>
<span><span class="va">project_rawdata</span> <span class="op">&lt;-</span> <span class="st">"/home/rstudio/macpie/macpieData/PMMSq033/raw_matrix/"</span></span>
<span><span class="va">raw_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X</a></span><span class="op">(</span>data.dir <span class="op">=</span> <span class="va">project_rawdata</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create tidySeurat object</span></span>
<span><span class="va">mac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">raw_counts</span>,</span>
<span>                          project <span class="op">=</span> <span class="va">project_name</span>,</span>
<span>                          min.cells <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                          min.features <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Feature names cannot have underscores ('_'), replacing with dashes</span></span>
<span><span class="co">#&gt; ('-')</span></span>
<span></span>
<span><span class="co"># Join with metadata</span></span>
<span><span class="va">mac</span> <span class="op">&lt;-</span> <span class="va">mac</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">inner_join</span><span class="op">(</span><span class="va">metadata</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">".cell"</span> <span class="op">=</span> <span class="st">"Barcode"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add unique identifier</span></span>
<span><span class="va">mac</span> <span class="op">&lt;-</span> <span class="va">mac</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>combined_id <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_c.html" class="external-link">str_c</a></span><span class="op">(</span><span class="va">Treatment_1</span>, <span class="va">Concentration_1</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>combined_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">" "</span>, <span class="st">""</span>, <span class="va">.data</span><span class="op">$</span><span class="va">combined_id</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Filter by read count per sample group</span></span>
<span><span class="va">mac</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_genes_by_expression.html">filter_genes_by_expression</a></span><span class="op">(</span><span class="va">mac</span>, </span>
<span>                                  group_by <span class="op">=</span> <span class="st">"combined_id"</span>, </span>
<span>                                  min_counts <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                                  min_samples <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate percent of mitochondrial and ribosomal genes</span></span>
<span><span class="va">mac</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">mac</span>, pattern <span class="op">=</span> <span class="st">"^mt-|^MT-"</span><span class="op">)</span></span>
<span><span class="va">mac</span><span class="op">[[</span><span class="st">"percent.ribo"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">mac</span>, pattern <span class="op">=</span> <span class="st">"^Rp[slp][[:digit:]]|^Rpsa|^RP[SLP][[:digit:]]|^RPSA"</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="single-comparison">3.1. Single comparison<a class="anchor" aria-label="anchor" href="#single-comparison"></a>
</h4>
<p>Similar to scRNA-seq data, MAC-seq gene expression counts have an
excess of zero counts compared to bulk RNA-seq. Statistical models
assuming a Poisson or negative binomial distribution may not fit the
data well. For that reason, differential expression effects could be
over or underestimated.</p>
<p>Let’s first perform the differential expression analysis with a
couple of methods and visualise the results on a volcano plot.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First perform the differential expression analysis</span></span>
<span><span class="va">treatment_samples</span> <span class="op">&lt;-</span> <span class="st">"Staurosporine_10"</span></span>
<span><span class="va">control_samples</span> <span class="op">&lt;-</span> <span class="st">"DMSO_0"</span></span>
<span><span class="va">top_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_single_de.html">compute_single_de</a></span><span class="op">(</span><span class="va">mac</span>, <span class="va">treatment_samples</span>, <span class="va">control_samples</span>, method <span class="op">=</span> <span class="st">"limma_voom"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Let's visualise the results with a volcano plot</span></span>
<span><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">top_table</span>, max.overlaps <span class="op">=</span> <span class="fl">18</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: ggrepel: 2947 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">#&gt; increasing max.overlaps</span></span></code></pre></div>
<p><img src="differential_expression_files/figure-html/de_analysis-1.png" width="768"></p>
<p>Based on the results, we can quickly check gene expression levels in
counts per million (CPM) for selected genes between treatment and
control samples as described below.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="va">top_table</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="va">group_by</span> <span class="op">&lt;-</span> <span class="st">"combined_id"</span></span>
<span><span class="fu"><a href="../reference/plot_counts.html">plot_counts</a></span><span class="op">(</span><span class="va">mac</span>,<span class="va">genes</span>, <span class="va">group_by</span>, <span class="va">treatment_samples</span>, <span class="va">control_samples</span>, normalisation <span class="op">=</span> <span class="st">"cpm"</span>, color_by <span class="op">=</span> <span class="st">"combined_id"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Normalizing layer: counts</span></span></code></pre></div>
<p><img src="differential_expression_files/figure-html/plot_counts-1.png" width="768"></p>
<p>Some plotting functions also have a “summarise” version that provides
collapsed versions of the results in a table format.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarise_de.html">summarise_de</a></span><span class="op">(</span><span class="va">top_table</span>, lfc_threshold <span class="op">=</span> <span class="fl">1</span>, padj_threshold <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 6</span></span></span>
<span><span class="co">#&gt;   Total_genes_tested Significantly_upregulated Significantly_downregulated</span></span>
<span><span class="co">#&gt;                <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>                     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>                       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>              <span style="text-decoration: underline;">20</span>114                       835                        <span style="text-decoration: underline;">2</span>136</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3 more variables: Total_significant &lt;int&gt;, Padj_threshold &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   Log2FC_threshold &lt;dbl&gt;</span></span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="pathway-analysis">3.2. Pathway analysis<a class="anchor" aria-label="anchor" href="#pathway-analysis"></a>
</h4>
<p>Differential gene expression results for individual comparisons of
treatment vs control can be easily performed with functions from package
<i>enrichR</i> and <i>fgsea</i>. In the following case, the effect of
Staurosporine on breast cancer cells through Myc inactivation can be
observed through pathway enrichment analyses. If you check the data from
“DisGeNET”, you will see that our mcf7 (breast cancer cell line) samples
are correctly enriched for breast cancer profiles.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_genes</span> <span class="op">&lt;-</span> <span class="va">top_table</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">p_value_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">pull</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">enriched</span> <span class="op">&lt;-</span> <span class="fu">enrichR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichR/man/enrichr.html" class="external-link">enrichr</a></span><span class="op">(</span><span class="va">top_genes</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MSigDB_Hallmark_2020"</span>,<span class="st">"DisGeNET"</span>,</span>
<span>                                 <span class="st">"RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Uploading data to Enrichr... Done.</span></span>
<span><span class="co">#&gt;   Querying MSigDB_Hallmark_2020... Done.</span></span>
<span><span class="co">#&gt;   Querying DisGeNET... Done.</span></span>
<span><span class="co">#&gt;   Querying RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO... Done.</span></span>
<span><span class="co">#&gt; Parsing results... Done.</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">enrichR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichR/man/plotEnrich.html" class="external-link">plotEnrich</a></span><span class="op">(</span><span class="va">enriched</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/macpie_theme.html">macpie_theme</a></span><span class="op">(</span>legend_position_ <span class="op">=</span> <span class="st">'right'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="va">macpie_colours</span><span class="op">$</span><span class="va">divergent</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span>
<span></span>
<span><span class="fu">gridExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span><span class="va">p1</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="differential_expression_files/figure-html/pathway_analysis_single-1.png" width="768"></p>
</div>
<div class="section level4">
<h4 id="differential-gene-expression---multiple-comparisons">3.3. Differential gene expression - multiple comparisons<a class="anchor" aria-label="anchor" href="#differential-gene-expression---multiple-comparisons"></a>
</h4>
<p>Since MAC-seq is commonly used for high-throughput screening of
compound libraries, we often want to compare multiple samples in a
screen vs the control. This process can easily be parallelised. First we
select a vector of “treatments” as combined_ids that do not contain the
word “DMSO”. (Warning, due to the limitations of “mclapply”,
parallelisation speedup currently only works on OSX and Linux machines,
and not on Windows.)</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mac</span><span class="op">$</span><span class="va">combined_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/make.names.html" class="external-link">make.names</a></span><span class="op">(</span><span class="va">mac</span><span class="op">$</span><span class="va">combined_id</span><span class="op">)</span></span>
<span></span>
<span><span class="va">treatments</span> <span class="op">&lt;-</span> <span class="va">mac</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Concentration_1</span> <span class="op">==</span> <span class="fl">10</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">combined_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"DMSO"</span>, <span class="va">combined_id</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">pull</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; tidyseurat says: Key columns are missing. A data frame is returned for independent data analysis.</span></span>
<span><span class="va">mac</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_multi_de.html">compute_multi_de</a></span><span class="op">(</span><span class="va">mac</span>, <span class="va">treatments</span>, control_samples <span class="op">=</span> <span class="st">"DMSO_0"</span>, method <span class="op">=</span> <span class="st">"limma_voom"</span>, num_cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>We often want to ask which genes are differentially expressed in more
than one treatment group.</p>
<p>Here, we can visualise treatment groups with shared differentially
expressed genes, defined as the top 5 DE genes from each single drug
comparison (treatment vs control) that are found in at least 2 different
treatment groups.</p>
<p>The heatmap below shows shared differentially expressed genes with
corresponding log2FC values.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_multi_de.html">plot_multi_de</a></span><span class="op">(</span><span class="va">mac</span>, group_by <span class="op">=</span> <span class="st">"combined_id"</span>, value <span class="op">=</span> <span class="st">"log2FC"</span>, p_value_cutoff <span class="op">=</span> <span class="fl">0.01</span>, direction<span class="op">=</span><span class="st">"up"</span>, n_genes <span class="op">=</span> <span class="fl">5</span>, control <span class="op">=</span> <span class="st">"DMSO_0"</span>, by<span class="op">=</span><span class="st">"fc"</span><span class="op">)</span></span></code></pre></div>
<p><img src="differential_expression_files/figure-html/plot_multi_de-1.png" width="960"></p>
<p>The outputs from the analyses above can be represented in table
format.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarise_de.html">summarise_de</a></span><span class="op">(</span><span class="va">mac</span>, lfc_threshold <span class="op">=</span> <span class="fl">1</span>, padj_threshold <span class="op">=</span> <span class="fl">0.01</span>, multi<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 27 × 7</span></span></span>
<span><span class="co">#&gt;    combined_id  Total_genes_tested Significantly_upregu…¹ Significantly_downre…²</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> Abemaciclib…              <span style="text-decoration: underline;">20</span>114                     36                     46</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Adavosertib…              <span style="text-decoration: underline;">20</span>114                    307                    104</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> Anastrozole…              <span style="text-decoration: underline;">20</span>114                      0                      1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> Azd.5991_10               <span style="text-decoration: underline;">20</span>114                      0                      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Camptotheci…              <span style="text-decoration: underline;">20</span>114                   <span style="text-decoration: underline;">1</span>108                   <span style="text-decoration: underline;">1</span>327</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> Capivaserti…              <span style="text-decoration: underline;">20</span>114                     22                     14</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> Ceralaserti…              <span style="text-decoration: underline;">20</span>114                    174                     88</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Chlorambuci…              <span style="text-decoration: underline;">20</span>114                      0                      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> Cytarabine_…              <span style="text-decoration: underline;">20</span>114                      8                      6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Erlotinib_h…              <span style="text-decoration: underline;">20</span>114                    252                    155</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 17 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ abbreviated names: ¹​Significantly_upregulated, ²​Significantly_downregulated</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3 more variables: Total_significant &lt;int&gt;, padj_threshold &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   Log2FC_threshold &lt;dbl&gt;</span></span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="pathway-analysis---multiple-comparisons">3.4. Pathway analysis - multiple comparisons<a class="anchor" aria-label="anchor" href="#pathway-analysis---multiple-comparisons"></a>
</h4>
<p>The pathway enrichment analysis is done by using enrichR</p>
<p>Summarise the outputs from multi de comparison in a table</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Load genesets from enrichr for a specific species or define your own</span></span>
<span><span class="va">enrichr_genesets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_geneset.html">download_geneset</a></span><span class="op">(</span><span class="st">"human"</span>, <span class="st">"MSigDB_Hallmark_2020"</span><span class="op">)</span></span>
<span><span class="va">mac</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_multi_enrichr.html">compute_multi_enrichr</a></span><span class="op">(</span><span class="va">mac</span>, genesets <span class="op">=</span> <span class="va">enrichr_genesets</span><span class="op">)</span></span>
<span></span>
<span><span class="va">enriched_pathways_mat</span> <span class="op">&lt;-</span> <span class="va">mac</span><span class="op">@</span><span class="va">tools</span><span class="op">$</span><span class="va">pathway_enrichment</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">bind_rows</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">combined_id</span>, <span class="va">Term</span>, <span class="va">Combined.Score</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">combined_id</span>, values_from <span class="op">=</span> <span class="va">Combined.Score</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"Term"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span><span class="fu">across</span><span class="op">(</span><span class="fu">everything</span><span class="op">(</span><span class="op">)</span>, <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span>, <span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>  <span class="co"># Replace NA with 0 across all columns</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">enriched_pathways_mat</span>, color <span class="op">=</span> <span class="va">macpie_colours</span><span class="op">$</span><span class="va">continuous_rev</span><span class="op">)</span></span></code></pre></div>
<p><img src="differential_expression_files/figure-html/enriched_pathways-1.png" width="768"></p>
<p><strong>Quick check of some treatments</strong>:</p>
<p><strong>Nutlin.3a</strong> is a MDM2-P53 inhibitor and stablises the
p53 protein. It induces cell autophagy and apotopsis. Nutlin-activated
p53 induces G1 and G2 arrest in cancer cell lines (see in the pathway
enrichment heatmap).</p>
<p><strong>Ref</strong>: Tovar C, et al. Proc Natl Acad Sci USA.
2006;103(6):1888–1893. Shows Nutlin-3’s effect on various p53 targets in
cancer cell lines.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nenad Bartonicek, Xin Liu, Laura Twomey, Michelle Meier, Richard Lupat, Stuart Craig, David Yoannidis, Jason Li, Tim Semple, Kaylene J Simpson, Mark X Li, Susanne Ramm.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
