<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • macpie</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://github.com/PMCC-BioinformaticsCore/macpie.git/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">macpie</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html">Home</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/metadata_check.html">Metadata check</a></li>
    <li><a class="dropdown-item" href="../articles/quality_control.html">Quality control</a></li>
    <li><a class="dropdown-item" href="../articles/differential_expression.html">Transcriptional analyses</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Compound screening</h6></li>
    <li><a class="dropdown-item" href="../articles/macpie_conversion.html">Interoperability</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="https://github.com/PMCC-BioinformaticsCore/macpie.git"><span class="fa fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="compute_de_umap.html">compute_de_umap()</a></code>

        </dt>
        <dd>Prepare DE-based umap</dd>
      </dl><dl><dt>

          <code><a href="compute_hyper_enrich_bg.html">compute_hyper_enrich_bg()</a></code>

        </dt>
        <dd>Calculate enrichment of DE genes in a gene set Internal function</dd>
      </dl><dl><dt>

          <code><a href="compute_multi_de.html">compute_multi_de()</a></code>

        </dt>
        <dd>Perform DE of multiple samples in a screen vs control</dd>
      </dl><dl><dt>

          <code><a href="compute_multi_enrichr.html">compute_multi_enrichr()</a></code>

        </dt>
        <dd>Perform enrichR-style analysis on a screen</dd>
      </dl><dl><dt>

          <code><a href="compute_multi_screen_profile.html">compute_multi_screen_profile()</a></code>

        </dt>
        <dd>Find similarities between expression profiles with fgsea</dd>
      </dl><dl><dt>

          <code><a href="compute_normalised_counts.html">compute_normalised_counts()</a></code>

        </dt>
        <dd>Retrieve normalised counts of MAC-seq data</dd>
      </dl><dl><dt>

          <code><a href="compute_qc_metrics.html">compute_qc_metrics()</a></code>

        </dt>
        <dd>Calculate QC metrics</dd>
      </dl><dl><dt>

          <code><a href="compute_single_de.html">compute_single_de()</a></code>

        </dt>
        <dd>Retrieve normalised counts of MAC-seq data</dd>
      </dl><dl><dt>

          <code><a href="compute_single_dose_response.html">compute_single_dose_response()</a></code>

        </dt>
        <dd>Model Gene Dose-Response Curve Using drc</dd>
      </dl><dl><dt>

          <code><a href="compute_single_enrichr.html">compute_single_enrichr()</a></code>

        </dt>
        <dd>Pathway enrichment analysis</dd>
      </dl><dl><dt>

          <code><a href="convert_human_to_mouse.html">convert_human_to_mouse()</a></code>

        </dt>
        <dd>For a given gene set of human symbols return mouse symbols</dd>
      </dl><dl><dt>

          <code><a href="download_geneset.html">download_geneset()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="filter_genes_by_expression.html">filter_genes_by_expression()</a></code>

        </dt>
        <dd>Filter genes by expression and grouping</dd>
      </dl><dl><dt>

          <code><a href="find_clusters_de_umap.html">find_clusters_de_umap()</a></code>

        </dt>
        <dd>Calculate clusters for umap based on DE analysis</dd>
      </dl><dl><dt>

          <code><a href="genesets.html">genesets</a></code>

        </dt>
        <dd>Pathway gene‐sets for PMMSq033</dd>
      </dl><dl><dt>

          <code><a href="macpie_colours.html">macpie_colours</a></code>

        </dt>
        <dd>Color palettes used by macpie</dd>
      </dl><dl><dt>

          <code><a href="macpie_theme.html">macpie_theme()</a></code>

        </dt>
        <dd>colour theme for macpie plots</dd>
      </dl><dl><dt>

          <code><a href="mini_mac.html">mini_mac</a></code>

        </dt>
        <dd>A small example MACseq plate</dd>
      </dl><dl><dt>

          <code><a href="plot_counts.html">plot_counts()</a></code>

        </dt>
        <dd>Generate a box plot to show gene expression (CPM) This is the function to generate a box plot to show CPM levels of DE genes among selected treatment samples and control samples.</dd>
      </dl><dl><dt>

          <code><a href="plot_de_umap.html">plot_de_umap()</a></code>

        </dt>
        <dd>Plot UMAP dimensionality reduction on DE genes</dd>
      </dl><dl><dt>

          <code><a href="plot_distance.html">plot_distance()</a></code>

        </dt>
        <dd>Create a distance heatmap</dd>
      </dl><dl><dt>

          <code><a href="plot_gene_ranks.html">plot_gene_ranks()</a></code>

        </dt>
        <dd>Generate a knee plot A knee plot to show total number of total read counts for each gene in a given treatment group</dd>
      </dl><dl><dt>

          <code><a href="plot_mds.html">plot_mds()</a></code>

        </dt>
        <dd>Plot MDS dimensionality reduction</dd>
      </dl><dl><dt>

          <code><a href="plot_metadata_heatmap.html">plot_metadata_heatmap()</a></code>

        </dt>
        <dd>Generate Heatmaps of Metadata Function</dd>
      </dl><dl><dt>

          <code><a href="plot_multi_de.html">plot_multi_de()</a></code>

        </dt>
        <dd>Generate heatmap of DE genes from multiple treatments This is the function to generate a heatmap of DE genes from running compute_multi_DE that shared by more than one treatment group. There are a few options available to help you to extract shared DE genes.</dd>
      </dl><dl><dt>

          <code><a href="plot_plate_layout.html">plot_plate_layout()</a></code>

        </dt>
        <dd>Plot MAC-seq data on a plate layout</dd>
      </dl><dl><dt>

          <code><a href="plot_qc_metrics.html">plot_qc_metrics()</a></code>

        </dt>
        <dd>Create a lollipop chart</dd>
      </dl><dl><dt>

          <code><a href="plot_qc_metrics_heatmap.html">plot_qc_metrics_heatmap()</a></code>

        </dt>
        <dd>Create a heatmap for multiple QC metrics</dd>
      </dl><dl><dt>

          <code><a href="plot_rle.html">plot_rle()</a></code>

        </dt>
        <dd>Create an RLE Plot</dd>
      </dl><dl><dt>

          <code><a href="plot_volcano.html">plot_volcano()</a></code>

        </dt>
        <dd>Volcano plot of differentially expressed genes</dd>
      </dl><dl><dt>

          <code><a href="read_metadata.html">read_metadata()</a></code>

        </dt>
        <dd>Read Metadata from a File</dd>
      </dl><dl><dt>

          <code><a href="summarise_de.html">summarise_de()</a></code>

        </dt>
        <dd>Generate a table to summarise gene numbers from a differential expression test.</dd>
      </dl><dl><dt>

          <code><a href="validate_metadata.html">validate_metadata()</a></code>

        </dt>
        <dd>Check and clean the metadata file generated from findmetadata function</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nenad Bartonicek, Xin Liu, Laura Twomey, Michelle Meier, Richard Lupat, Stuart Craig, David Yoannidis, Jason Li, Tim Semple, Kaylene J Simpson, Mark X Li, Susanne Ramm.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

