<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Data integration ‚Ä¢ macpie</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="‚Äùimage/svg+xml‚Äù" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Data integration">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">macpie</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html">Home</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/macpie.html">Overview</a></li>
    <li><a class="dropdown-item" href="../articles/quality_control.html">Quality control</a></li>
    <li><a class="dropdown-item" href="../articles/transcriptional_analyses.html">Transcriptional analyses</a></li>
    <li><a class="dropdown-item" href="../articles/high_throughput_screens.html">Compound screening</a></li>
    <li><a class="dropdown-item" href="../articles/cross_platform_compatibility.html">Cross-platform compatibility</a></li>
    <li><a class="dropdown-item" href="../articles/integration_across_modalities.html">Integration across modalities</a></li>
    <li><a class="dropdown-item" href="../articles/macpie_bioc.html">Working with Bioconductor classes</a></li>
    <li><a class="dropdown-item" href="../articles/assessing_zero_inflation.html">Assessing zero-inflation</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="https://github.com/PMCC-BioinformaticsCore/macpie.git"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Data integration</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/PMCC-BioinformaticsCore/macpie/blob/HEAD/vignettes/integration_across_modalities.Rmd" class="external-link"><code>vignettes/integration_across_modalities.Rmd</code></a></small>
      <div class="d-none name"><code>integration_across_modalities.Rmd</code></div>
    </div>

    
    
<p>High throughput transcriptomic screens present a challenge of
combining information across modalities to identify patterns and
relationships. Here we present an approaches using unsupervised variance
decomposition with Multi-Omics Factor Analysis (<a href="https://biofam.github.io/MOFA2/" class="external-link">MOFA</a>).</p>
<p>In this vignette we show how to integrate multiple data modalities,
including gene expression, pathway enrichment, chemical descriptors, and
phenotypic readouts such as cell viability. While the example dataset
itself lacks power to build stable models, we wanted to demonstrate the
basic functionality, using the macpie framework.</p>
<div class="section level3">
<h3 id="import-screen-data">1. Import screen data<a class="anchor" aria-label="anchor" href="#import-screen-data"></a>
</h3>
<p>First import data by providing either a directory, or a vector of
directories (for multiple plates) to the Read10X function, as described
in the previous vignettes, such as <a href="quality_control.html">Quality control</a>.</p>
</div>
<div class="section level3">
<h3 id="data-collection">2. Data collection<a class="anchor" aria-label="anchor" href="#data-collection"></a>
</h3>
<p>In this section we will collate data from across various modalities
regularly associated with transcriptomic screens. We will show how to
import sample annotations, and integrate them with gene expression and
pathway enrichment.</p>
<div class="section level4">
<h4 id="import-external-annotations">2.1 Import external annotations<a class="anchor" aria-label="anchor" href="#import-external-annotations"></a>
</h4>
<p>Below is an example of loading additional sample information, in this
case cell viability and confluence, and merging it with metadata based
on the column Well_ID.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"PMMSq033/PMMSq033_CTG_cellcount.csv"</span>, package <span class="op">=</span> <span class="st">"macpie"</span><span class="op">)</span></span>
<span><span class="va">cell_viability</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">file_path</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co">#mutate(</span></span>
<span>  <span class="co">#  cell_viability = scale(cell_viability)[, 1],</span></span>
<span>  <span class="co">#  cell_confluence = scale(cell_confluence)[, 1]</span></span>
<span>  <span class="co">#) %&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Well_ID</span>, <span class="va">cell_viability</span>, <span class="va">cell_confluence</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Modify metadata</span></span>
<span><span class="va">mac</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="va">mac</span><span class="op">@</span><span class="va">meta.data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">left_join</span><span class="op">(</span><span class="va">cell_viability</span>, by <span class="op">=</span> <span class="st">"Well_ID"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span><span class="st">"Barcode"</span><span class="op">)</span> <span class="co">#tidyverse does not preserve row names that Seurat needs</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="calculate-differentially-expressed-genes">2.2 Calculate differentially expressed genes<a class="anchor" aria-label="anchor" href="#calculate-differentially-expressed-genes"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">treatments</span> <span class="op">&lt;-</span> <span class="va">mac</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Concentration_1</span> <span class="op">==</span> <span class="fl">10</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">combined_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"DMSO"</span>, <span class="va">combined_id</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">pull</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">mac</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_multi_de.html">compute_multi_de</a></span><span class="op">(</span><span class="va">mac</span>, <span class="va">treatments</span>, control_samples <span class="op">=</span> <span class="st">"DMSO_0"</span>, method <span class="op">=</span> <span class="st">"limma_voom"</span>, num_cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="calculate-pathway-enrichment">2.3 Calculate pathway enrichment<a class="anchor" aria-label="anchor" href="#calculate-pathway-enrichment"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrichr_genesets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_geneset.html">download_geneset</a></span><span class="op">(</span><span class="st">"human"</span>, <span class="st">"MSigDB_Hallmark_2020"</span><span class="op">)</span></span>
<span><span class="va">mac</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_multi_enrichr.html">compute_multi_enrichr</a></span><span class="op">(</span><span class="va">mac</span>, genesets <span class="op">=</span> <span class="va">enrichr_genesets</span>, p_value_cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="calculate-chemical-descriptors-of-compounds">2.4 Calculate chemical descriptors of compounds<a class="anchor" aria-label="anchor" href="#calculate-chemical-descriptors-of-compounds"></a>
</h4>
<p>If your perturbations are small molecules, it is straightforward to
add information on chemical descriptors based on the smiles column in
your data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># If you don't have smiles in your metadata, calculate them based on the compound ID </span></span>
<span><span class="co">#mac &lt;- compute_smiles(mac, compound_column = "Compound_ID")</span></span>
<span><span class="va">mac</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_chem_descriptors.html">compute_chem_descriptors</a></span><span class="op">(</span><span class="va">mac</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="mofa-workflow">3. MOFA workflow<a class="anchor" aria-label="anchor" href="#mofa-workflow"></a>
</h3>
<p>MOFA provides a framework for the integration of multiomics datasets,
by identifying ‚ÄúFactors‚Äù (principal components) that define main sources
of variation in the dataset, across the modalities. In order to identify
them, we will follow the MOFA workflow.</p>
<p>The key functions include:</p>
<ul>
<li>create_mofa (import multiomics data.frame in long format)</li>
<li>prepare_mofa (set traning options)</li>
<li>run_mofa (train the MOFA model)</li>
</ul>
<p>The <strong>macpie</strong> package streamlines analysis with MOFA,
by extracting all multimodal data that is required for the first step
and creating the MOFA object for the user.</p>
<p>To load data into a ‚ÄúMOFAobject‚Äù and start the analysis, we us
<strong>build_mofa</strong>, a function that prepares a MOFA object
based on</p>
<ul>
<li>genes and pathways (previously calculated, in slot <span class="citation">@tools</span>)</li>
<li>chemical descriptors (previously calculated, optional)</li>
<li>any other tables in the slot <span class="citation">@tools</span>
containing a column ‚ÄúTreatment‚Äù, with unique treatment ids</li>
</ul>
<p>In the first instance we will try to identify Factors that capture
variation in both chemical descriptors and gene expression / pathway
enrichment.</p>
<div class="section level4">
<h4 id="mofa-object">3.1 MOFA object<a class="anchor" aria-label="anchor" href="#mofa-object"></a>
</h4>
<p>We include metadata columns that define cell viability and
confluence.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MOFAobject</span> <span class="op">&lt;-</span> <span class="fu">build_mofa</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mac</span>,</span>
<span>                         combined_ids <span class="op">=</span> <span class="va">treatments</span>, <span class="co">#unique samples IDs (treat+conc) to be used</span></span>
<span>                         metadata_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell_confluence"</span>, <span class="st">"cell_viability"</span><span class="op">)</span>, <span class="co">#vector of metadata columns to include in the analysis</span></span>
<span>                         de_metric <span class="op">=</span> <span class="st">"metric"</span>, <span class="co">#column name of data for DE analysis, such as: </span></span>
<span>                                              <span class="co">#"log2FC" or "metric" (test statistic)</span></span>
<span>                         de_pval_thresh <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>                         de_padj_col <span class="op">=</span> <span class="st">"p_value_adj"</span>,</span>
<span>                         pathway_pval_thresh <span class="op">=</span> <span class="fl">1</span>, <span class="co">#p-value cutoff for pathway significance</span></span>
<span>                         pathway_pval_col <span class="op">=</span> <span class="st">"Adjusted.P.value"</span>, <span class="co">#column used for p-valye cutoff</span></span>
<span>                         pathway_score_col <span class="op">=</span> <span class="st">"Combined.Score"</span> <span class="co">#column used as pathway enrichment metric</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">MOFA2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MOFA2/man/plot_data_overview.html" class="external-link">plot_data_overview</a></span><span class="op">(</span><span class="va">MOFAobject</span><span class="op">)</span></span></code></pre></div>
<p><img src="integration_across_modalities_files/figure-html/prepare_mofa-1.png" width="768"></p>
</div>
<div class="section level4">
<h4 id="scaling">3.2 Scaling<a class="anchor" aria-label="anchor" href="#scaling"></a>
</h4>
<p>If groups or modalities (views) have different ranges/variances, it
is good practice to scale them to unit variance by modifying the model
options scale_views and scale_groups (default is FALSE).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_opts</span> <span class="op">&lt;-</span> <span class="fu">MOFA2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MOFA2/man/get_default_data_options.html" class="external-link">get_default_data_options</a></span><span class="op">(</span><span class="va">MOFAobject</span><span class="op">)</span></span>
<span><span class="va">data_opts</span><span class="op">$</span><span class="va">scale_views</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span><span class="va">data_opts</span><span class="op">$</span><span class="va">scale_groups</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="model-parameters">3.3 Model parameters<a class="anchor" aria-label="anchor" href="#model-parameters"></a>
</h4>
<p>Modify model and training parameters here, including likelihoods of
distributions, number of iterations and usage of GPU.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_opts</span> <span class="op">&lt;-</span> <span class="fu">MOFA2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MOFA2/man/get_default_model_options.html" class="external-link">get_default_model_options</a></span><span class="op">(</span><span class="va">MOFAobject</span><span class="op">)</span></span>
<span><span class="va">train_opts</span> <span class="op">&lt;-</span> <span class="fu">MOFA2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MOFA2/man/get_default_training_options.html" class="external-link">get_default_training_options</a></span><span class="op">(</span><span class="va">MOFAobject</span><span class="op">)</span></span>
<span><span class="va">train_opts</span><span class="op">$</span><span class="va">seed</span> <span class="op">&lt;-</span> <span class="fl">1</span>  <span class="co"># for reproducibility</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="running-the-model">3.4 Running the model<a class="anchor" aria-label="anchor" href="#running-the-model"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MOFAobject</span> <span class="op">&lt;-</span> <span class="fu">MOFA2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MOFA2/man/prepare_mofa.html" class="external-link">prepare_mofa</a></span><span class="op">(</span><span class="va">MOFAobject</span>, </span>
<span>                           data_options <span class="op">=</span> <span class="va">data_opts</span>,</span>
<span>                           model_options <span class="op">=</span> <span class="va">model_opts</span>, </span>
<span>                           training_options <span class="op">=</span> <span class="va">train_opts</span><span class="op">)</span></span>
<span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu">MOFA2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MOFA2/man/run_mofa.html" class="external-link">run_mofa</a></span><span class="op">(</span><span class="va">MOFAobject</span>, use_basilisk <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="investigation-of-factors">3.5 Investigation of Factors<a class="anchor" aria-label="anchor" href="#investigation-of-factors"></a>
</h4>
<p>This plot allows us to evaluate whether the Factors are correlated,
in which case they should be eliminated from the future analyses.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">MOFA2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MOFA2/man/plot_factor_cor.html" class="external-link">plot_factor_cor</a></span><span class="op">(</span><span class="va">model</span><span class="op">)</span></span></code></pre></div>
<p><img src="integration_across_modalities_files/figure-html/mofa_factor_correlation-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="variance-decomposition">3.6 Variance decomposition<a class="anchor" aria-label="anchor" href="#variance-decomposition"></a>
</h4>
<p>After running the model, it is now easy to follow the standard MOFA
worklow. We can quantify the amount of variance explained (ùëÖ2) by each
factor, for each data modality.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">model</span><span class="op">@</span><span class="va">cache</span><span class="op">$</span><span class="va">variance_explained</span><span class="op">$</span><span class="va">r2_per_factor</span><span class="op">$</span><span class="va">group1</span><span class="op">)</span> <span class="co"># group 1</span></span>
<span><span class="co">#&gt;         cell_confluence cell_viability     genes     pathways chem_descriptors</span></span>
<span><span class="co">#&gt; Factor1    3.829609e+01   1.993528e+01 34.349394 0.0814557076                0</span></span>
<span><span class="co">#&gt; Factor2    7.075602e+00   3.376725e+01  5.242789 0.0003278255                0</span></span>
<span><span class="co">#&gt; Factor3    1.805512e+01   1.282811e-01  9.765500 0.0020205975                0</span></span>
<span><span class="co">#&gt; Factor4    4.338485e+00   1.340080e+01  3.388435 0.0007033348                0</span></span>
<span><span class="co">#&gt; Factor5   -4.768372e-05   4.529953e-04  3.786212 0.0009715557                0</span></span></code></pre></div>
<p>It seems that the first factor captures a lot of variability across
all modalities, with the least for pathways (8.1%) and chemical
descriptors (0%). This could be explained by the fact that a lot of
information content in pathways is directly contained in gene
expression, which is already captured by MOFA. Also, the number of
molecules might be too small for their chemical properties to be be able
to explain variability in the dataset.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">&lt;-</span><span class="fu">MOFA2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MOFA2/man/plot_variance_explained.html" class="external-link">plot_variance_explained</a></span><span class="op">(</span><span class="va">model</span><span class="op">)</span></span>
<span><span class="va">p</span><span class="op">&lt;-</span><span class="va">p</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="integration_across_modalities_files/figure-html/mofa_factors-1.png" width="768"></p>
</div>
<div class="section level4">
<h4 id="visualisation-of-feature-weights">3.7 Visualisation of feature weights<a class="anchor" aria-label="anchor" href="#visualisation-of-feature-weights"></a>
</h4>
<p>If ‚ÄúFactors‚Äù correspond to principle componenets of PCA analysis,
then ‚Äúweight‚Äù corresponds to ‚Äúloadings‚Äù. A positive weight for a
feature, in this case genes such as PPP2R1A and DCTN3, marks a strong
relation to Factor 1. The sign of the weight indicates the direction of
the effect, so features PPP2R1A and DCTN3 have higher levels in the
cells with positive Factor 1 values, while cell viability has the
opposite.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span><span class="op">&lt;-</span><span class="fu">MOFA2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MOFA2/man/plot_top_weights.html" class="external-link">plot_top_weights</a></span><span class="op">(</span><span class="va">model</span>,</span>
<span>                 view <span class="op">=</span> <span class="st">"genes"</span>,</span>
<span>                 factor <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                 nfeatures <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p2</span><span class="op">&lt;-</span><span class="fu">MOFA2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MOFA2/man/plot_top_weights.html" class="external-link">plot_top_weights</a></span><span class="op">(</span><span class="va">model</span>,</span>
<span>                 view <span class="op">=</span> <span class="st">"cell_viability"</span>,</span>
<span>                 factor <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                 nfeatures <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p1</span><span class="op">+</span><span class="va">p2</span></span></code></pre></div>
<p><img src="integration_across_modalities_files/figure-html/mofa_genes-1.png" width="768"></p>
<p>We can then use MOFA‚Äôs scatter plots to visualise factor values, and
establish whether relationship between features and factor values is
linear</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">plot_data_scatter</span><span class="op">(</span><span class="va">model</span>,</span>
<span>  view <span class="op">=</span> <span class="st">"genes"</span>,         <span class="co"># view of interest</span></span>
<span>  factor <span class="op">=</span> <span class="fl">1</span>,             <span class="co"># factor of interest</span></span>
<span>  features <span class="op">=</span> <span class="fl">5</span>,           <span class="co"># number of features to plot (they are selected by weight)</span></span>
<span>  add_lm <span class="op">=</span> <span class="cn">TRUE</span>,          <span class="co"># add linear regression</span></span>
<span>  color_by <span class="op">=</span> <span class="st">"cell_viability"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="integration_across_modalities_files/figure-html/mofa_genes_viability-1.png" width="768"></p>
<p>DCTN3 encodes dynactin subunit 3, a part of dynactin motor complex
essential for intracellular transport and formation of mitotic spindle.
High DCTN3 may influence cell viability in cells treated with
microtubule inhibitors. Similarly, PPP2R1 is a serine/threonine
phosphatase that regulates chromatin remodeling, and high levels of
PPP2R1A may be detrimental in combination with epigenetic drugs.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span><span class="op">&lt;-</span> <span class="fu">plot_factor</span><span class="op">(</span><span class="va">model</span>, </span>
<span>            factors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>,</span>
<span>            group_by <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>            color_by <span class="op">=</span> <span class="st">"DCTN3"</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, vjust <span class="op">=</span> <span class="fl">0.5</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#p1</span></span>
<span><span class="va">p2</span><span class="op">&lt;-</span> <span class="fu">plot_factor</span><span class="op">(</span><span class="va">model</span>, </span>
<span>            factors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>,</span>
<span>            group_by <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>            color_by <span class="op">=</span> <span class="st">"cell_viability"</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, vjust <span class="op">=</span> <span class="fl">0.5</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p1</span><span class="op">+</span><span class="va">p2</span></span></code></pre></div>
<p><img src="integration_across_modalities_files/figure-html/mofa_factors_individual-1.png" width="768"></p>
<p>DCTN3 expression appears to stratify differential cellular responses
to diverse therapeutic agents. High DCTN3 levels correlate with reduced
viability in cells treated with camptothecin, 5‚Äô-azacytidine, and
paclitaxel - drugs that rely on active cell division and efficient
mitotic machinery for their cytotoxic effects, suggesting that
DCTN3-dependent trafficking and spindle dynamics may sensitize cells to
genotoxic stress. Conversely, low DCTN3 expression is associated with
increased viability in response to capivasertib, cytarabine, ribociclib,
and SB590885, potentially reflecting a drug-tolerant state marked by
reduced proliferation, impaired retrograde transport, and altered
checkpoint engagement.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nenad Bartonicek, Xin Liu, Laura Twomey, Michelle Meier, Richard Lupat, Stuart Craig, David Yoannidis, Jason Li, Tim Semple, Kaylene J Simpson, Mark X Li, Susanne Ramm.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
